// @hoge1e3/immer-path-ref/js/proxy.js
import { ref } from "@hoge1e3/ref";
import { fromRef as c } from "./immer-path-ref.js";
export function map(p, m) {
    return fromR(p.__unproxy__.join(m));
}
export function fromRef(ref) {
    const b = c(ref);
    return fromW(b);
}
export function fromVal(val) {
    return fromRef(ref(val));
}
function getBind(target, p) {
    const r = Reflect.get(target, p);
    if (typeof r === "function")
        return r.bind(target);
    else
        return r;
}
function fromW(b) {
    return new Proxy(b, {
        get(target, p) {
            if (p === "__unproxy__") {
                return target;
            }
            if (typeof p === "symbol" || p === "then" || p === "toString") {
                return getBind(target, p);
            }
            return fromW(target.join(p));
        },
        set(target, p, v) {
            if (typeof p === "symbol") {
                Reflect.set(target, p, v);
                return true;
            }
            throw new Error(`Cannot set via ***.${p}. use set( ) instead `);
        },
    });
}
function fromR(b) {
    return new Proxy(b, {
        get(target, p) {
            if (p === "__unproxy__") {
                return target;
            }
            if (typeof p === "symbol" || p === "then" || p === "toString") {
                return getBind(target, p);
            }
            return fromW(target.join(p));
        },
        set(target, p, v) {
            if (typeof p === "symbol") {
                Reflect.set(target, p, v);
                return true;
            }
            throw new Error(`Cannot set via ***.${p}. use set( ) instead `);
        },
    });
}
