// ace-local-commands/src/ace-local-commands.ts

import type {
  edit, Editor, EditSession,
} from "@hoge1e3/ace-types";
import type { Ace } from "@hoge1e3/ace-types";

type CommandMap = Record<string, Ace.Command>;

const editor2globals = new WeakMap<Editor, CommandMap>();
const session2locals = new WeakMap<EditSession, CommandMap>();

export function init(editor: Editor) {
    if (editor2globals.has(editor)) return;
    getGlobalCommands(editor);
    editor.on("changeSession", ({ oldSession, session }) => {
//        console.log("change",oldSession,session);
        disableLocalCommands(editor, oldSession);
        restoreLocalCommands(editor, session);
    });
}
function getGlobalCommands(editor: Editor): CommandMap {
    if (!editor2globals.has(editor)) {
        editor2globals.set(editor, {});
    }
    return editor2globals.get(editor)!;
}
function getLocalCommands(session: EditSession): CommandMap {
    if (!session2locals.has(session)) {
        session2locals.set(session, {});
    }
    return session2locals.get(session)!;
}
export function addGlobalCommands(editor: Editor, commands: CommandMap) {
    init(editor);
    const globalCommands = getGlobalCommands(editor);
    Object.assign(globalCommands, commands);
    editor.commands.addCommands(commands);
}
export function removeGlobalCommands(editor: Editor, commands: CommandMap) {
    init(editor);
    disableLocalCommands(editor);
    const globalCommands = getGlobalCommands(editor);
    for (const k in commands) {
        delete globalCommands[k];
    }
    editor.commands.removeCommands(commands);
    restoreLocalCommands(editor);
}
export function addCommandsToSession(session: EditSession, commands: CommandMap) {
    const localCommands = getLocalCommands(session);
    Object.assign(localCommands, commands);
}
export function addLocalCommands(editor: Editor, commands: CommandMap) {
    init(editor);
    const session = editor.session;
    addCommandsToSession(session, commands);
    editor.commands.addCommands(commands);
}
export function removeLocalCommands(editor: Editor, commands: CommandMap) {
    init(editor);
    const session = editor.session;
    const localCommands = getLocalCommands(session);
    for (const k in commands) {
        delete localCommands[k];
    }
    editor.commands.removeCommands(commands);
}
function disableLocalCommands(editor: Editor, session: EditSession = editor.session) {
    const localCommands = getLocalCommands(session);
    editor.commands.removeCommands(localCommands);
    restoreGlobalCommands(editor);
}
function restoreGlobalCommands(editor: Editor) {
    const globalCommands = getGlobalCommands(editor);
    editor.commands.addCommands(globalCommands);
}
function restoreLocalCommands(editor: Editor, session: EditSession = editor.session) {
    const localCommands = getLocalCommands(session);
    editor.commands.addCommands(localCommands);
}
