// @acepad/worker/acepad-worker.js
// acepad-worker
import * as npm from "@acepad/npm";
import * as sw from "@hoge1e3/str2worker";
import * as rpc from 
    "https://unpkg.com/@hoge1e3/rpc@latest/dist/index.js";
import * as cons from "./console.js";
import {sibling} from "@acepad/here";
export * as cons from "./console.js";
const em=(d)=>JSON.stringify(d);
async function createFixture(rpa){
    const deps=await npm.getDeps(rpa,true);
    deps.add(rpa);
    const data={};
    for(let pa of deps){
        pa=pa.up();
        for(let f of pa.recursive()){
            data[f.path()]=f.text();
        }
    }
    return data;
}
const cdn="https://cdn.jsdelivr.net/npm/";//"https://unpkg.com/"
function header(){
    let purl=process.env.PNODE_URL;
    // static import does not reports error
    // using relative path for import does not work in BlobURL?
    if (!purl || !purl.match(/^https:\/\//)) purl=cdn+"petit-node@latest/dist/index.js";
    return `const pNode=await import(${em(purl)});
function loadFixture(FS,fixture) {
    for (let name in fixture) {
        const f=FS.get(name);
        f.text(fixture[name]);
    }
}
`;
}
function rpcHeader(){
  const rurl=`${cdn}@hoge1e3/rpc@latest/dist/index.js`;
  return `const rpc=await import (${em(rurl)});
`;
}
function hackTimeouts(){
    for(let k of ["setTimeout","setInterval",
    "clearTimeout","clearInterval",]){
        globalThis[k]=globalThis[k].bind(globalThis);
    }
}

export async function createProxy(main, rev={}){
    /*const rpa=npm.findPackageJson(f);
    const main=rpa.sibling(rpa.obj().main);*/
    const fixture={};//await createFixture(rpa);
    const consolePath=sibling(
        import.meta.url,"console.js").
        resolveLink().path();
    const ch_rev=Math.random()+"";
    const src=`
try{
${rpcHeader()}
${header()}
(${hackTimeouts})();
const fixture=${em(fixture)};
await pNode.boot({
    async init({FS}){
        try{
            //loadFixture(FS, fixture);
            // Not exists when nw-js
            if (FS.mountAsync) await FS.mountAsync("/idb/","idb");
            const main=FS.get(${em(main.path())});
            const mod=await pNode.importModule(main);
            const srv={};
            const rev_cli=rpc.proxy.client(self, ${em(ch_rev)});
            for(let k in mod){
                let v=mod[k];
                if(typeof v==="function"){
                    srv[k]=v.bind(rev_cli);
                }
            }
            rpc.proxy.server("default",[],srv);
            const cons=await pNode.import(FS.get(${em(consolePath)}));
            cons.client();
        }catch(e){
            console.error(e);
            self.postMessage(e.stack);
            throw e;
        }
    },
});
}catch(e) {
  console.error(e);
}
`;
    console.log("worker-gen",src);
    const w=sw.create(src);
    console.log("generated",w);
    /*w.addEventListener("message",(r)=>{
        console.log("From worker",ch_rev,r);
    });*/
    cons.server(w);
    const revserv=rpc.proxy.server(w, ch_rev, [], rev);
    console.log("revserv",revserv);
    return rpc.proxy.client(w,"default");

    
}
export async function create(main){
    const src=`
${header()}
await pNode.boot({
    async init({FS}){
        return FS.get(${em(main.path())});
    },
});
`;
    console.log(src);
    const w=sw.create(src);
    return w;
    
}
