// @acepad/worker/from-bootloader.js
#!run
import {getDeviceManager} from "pnode:main";
import * as rpc from "@hoge1e3/rpc";
import * as cons from "./console.js";
import {file} from "@acepad/here";
import {init} from "@hoge1e3/workarea";
import isPlainObject from "@hoge1e3/is-plain-object";

function uid(m){
 let c=m.size;
  for(let i=0;i<100;i++){
   let id=`w${c}${i>90?Math.random():""}`;
   if(!m.has(id)){
     m.set(id,{});
     return id;
   }
   c++;
 }
}
const id2w=
  init("@acepad/worker",
  ()=>new Map());
export async function createInstance(main, opt={}){
  if (typeof main!=="string"){
    main=main.path();
  } else {
    if (typeof this.resolve==="function") {
      main=this.resolve(main).path();
    } else {
      main=file(main).path();
    }
  }
  const worker=new Worker(
    process.env.PNODE_BOOTLOADER_URL||
    process.env.WEBCARTRIDGE_MAIN,{
      type:"module"
    });
  cons.server(worker);
  const client=rpc.proxy.client(worker,"boot");
  const dev=getDeviceManager();
  const fstab=(
    typeof dev.currentFstab==="function"?
    dev.currentFstab().filter(t=>t.mountPoint!=="/"):
    dev.readFstab("/fstab.json",[]));
  const id=uid(id2w);
  await client.boot({
    fstab,
    id,
    main
  });
  const proxy=rpc.proxy.client(worker,"default");
  const revproxy=rpc.proxy.server(
    worker, id, [],
    opt.reverse||{});
  const instance={
    ...opt,
    id,
    worker,
    client,
    proxy,
    revproxy,
    main,
  };
  id2w.set(id,instance/*{
    id,
    ref:new WeakRef(instance),
    main,
  }*/);
  return instance;
}
export async function createProxy(main, reverse={}){
  return (await createInstance(main,{reverse})).proxy;
}
export async function createWorker(main){
  const instance=await createInstance(main);   
  return instance.worker;
}
export function kill(o){
  const m=meta(o);
  if(!m){
    return ;
  }
  const i=m;//.ref.deref();
  if(i)i.worker.terminate();
  id2w.delete(m.id);
  return true;
}
export function meta(o){
  if(!o)return id2w.values();
  if(typeof o==="string"){
    return id2w.get(o);
  }
  if(!isPlainObject(o)){
    o={
      proxy:o
    };
  }
  const eq=(a,b)=>{
    if(typeof a==="object"&&
    typeof a?.path==="function"){
      a=a.path();
    }
    if(typeof b==="object"&&
    typeof b?.path==="function"){
      b=b.path();
    }
    return a===b;
  }
  for(let [k,v] of id2w){
    if(Object.keys(o).every(
      (k)=>eq(v[k],o[k])))return v;
  }
}
function instanceProxy(c){
  //todo
  return new Proxy(c, {
        get(target, p) {
            if (p === "__unproxy__") {
                return c;
            }
            if (typeof p === "symbol" || p === "then" || p === "toString") {
                return Reflect.get(target, p);
            }
            
        }
    });
}
