// @hoge1e3/html-builder/html.js

export default function builder(target,...args){
    const b=Builder.clone({target});
    return b.apply(...args);
}
builder.custom=(customs)=>{
    const b=Object.create(Builder);
    for(let name in customs){
        let f=customs[name];
        b[name]=function (...args){
            return f(p(this),...args);
        };
    }
    return function (target,...args){
        return b.clone({target}).apply(...args);
    };
};    
function p(obj){
    const handler={
        get(_o,prop){
            if(prop in obj){
                const v=obj[prop];
                if(typeof v==="function"){
                    return v.bind(obj);
                }
                return v;
            }
            return function (...a){
                return obj.tag(prop,...a);
            };
        },
        apply(_,tx,args){
            return obj.apply(...args);    
        },
        set(_o,k,v){
            return Reflect.set(obj,k,v);
        },
    };
    return new Proxy(function (){},handler);    
}
const Builder={
    tag(name,...a){
        const nt=document.createElement(name);
        this.target.appendChild(nt);
        this.clone({target:nt}).apply(...a);
        return nt;
    },
    custom(name,f){
        this[name]=function (...args){
            return f(p(this),...args);
        };
    },
    apply1(e){
        const d=this.target;
        switch(typeof e){
            case "string":
            case "number":
            case "boolean":
                return this.applyPrimitive(e);
            case "function":
                return this.applyFunction(e);
            case "object":
                if(this.isTarget(e)){
                    return this.applyTarget(e);
                }else{
                    return this.applyAttributes(e);
                }
        }
    },
    apply(...a){
        while(a.length){
            const e=a.shift();
            this.apply1(e);
        }
    },
    applyAttribute(k,v){
        this.target.setAttribute(k,v);
    },
    applyAttributes(a){
        for(let k in a){
            this.applyAttribute(k,a[k]);
        }
    },
    applyTarget(t){
        this.target.appendChild(t);
    },
    isTarget(t){
        return t instanceof Node;        
    },
    applyPrimitive(s){
        const n=document.createTextNode(s+"");
        this.target.appendChild(n);
    },
    applyFunction(f){
        return f(p(this));
    },
    clone(param){
        const b=Object.create(this);
        Object.assign(b,param);
        return b;
    },
};
