// @hoge1e3/immer-path-ref/src/immer-path-ref.ts
// @hoge1e3/immer-path-ref
import {produce} from "immer";
import { Ref,ReadonlyRef,custom,computed } from "@hoge1e3/ref";
export interface ReadonlyPathRef<T> extends ReadonlyRef<T> {
    <N>(path: ((t: T) => N)): ReadonlyPathRef<N>;
    <K extends keyof T>(path: K): PathRef<T[K]>;
    join<N>(path: ((t: T) => N)): ReadonlyPathRef<N>;
    join<K extends keyof T>(path: K): PathRef<T[K]>;
}
export interface PathRef<T> extends Ref<T> {
    <N>(path: ((t: T) => N)): ReadonlyPathRef<N>;
    <K extends keyof T>(path: K): PathRef<T[K]>;
    join<N>(path: ((t: T) => N)): ReadonlyPathRef<N>;
    join<K extends keyof T>(path: K): PathRef<T[K]>;
}
export type PathComponent<T, N> =
  | ((t: T) => N)
  | (N extends T[keyof T]
      ? { [K in keyof T]: T[K] extends N ? K : never }[keyof T]
      : never);
export type Path<R, T> =
  | null
  | ConsPath<R, any, T>;
export interface ConsPath<R, M, T> {
  left: Path<R, M>;
  right: PathComponent<M, T>;
}
export function applyPath<R, T>(root:R,path:null):R;
export function applyPath<R, T>(
  root: R,
  path: Path<R, T>
):T;
export function applyPath<R, T>(
  root: R,
  path: Path<R, T>
): T {
  if (!path) return root as unknown as T;
  const { left, right } = path;
  const rv = applyPath(root, left);
  if (typeof right === "function") {
    return right(rv);
  } else {
    return rv[right];
  }
}
export function create<T>(
  r: Ref<T>,
): PathRef<T> {
  return joinable(r,null);
}
function joinable<R,T>(
  root:Ref<R>,
  path: Path<R,T>):PathRef<T>{
  let cur=applyPath(root.value,path)
  root.on("change",(e)=>{
    const newval=applyPath(e.val,path); 
    if (newval===cur) return;
    const prev=cur;
    cur=newval;
    res.fire("change",{prev,val:cur});
  });
  /*
  join<N>(path: ((t: T) => N)): ReadonlyPathRef<N>;
  join<K extends keyof T>(path: K): PathRef<T[K]>;
  */
  const join=<N>(right:PathComponent<T,N>):PathRef<N>=>{
    return joinable(root, {left: path, right:right as any});
  };
  const res= Object.assign(custom(
      ()=>cur,
      (value:T)=>{
        cur=value;
        if (path==null)root.value=value as any as R;
        else root.value=produce(root.value,(draft:R)=>{
          const lv=applyPath(draft, path.left);
          if (typeof path.right==="function") {
            throw new Error("Read only path!");
          } else {
            lv[path.right]=value;
          }
        });
      },
    ),Object.assign(join,{join}));
  return res;
}
