// @hoge1e3/immer-path-ref/src/proxy.ts
import { RefProxy,ref,Ref,set,
ReadonlyRefProxy,custom,computed } from "@hoge1e3/ref";
import {produce} from "immer";
import {fromRef as c,Path,
ReadonlyPathRef,applyPath,PathRef,PathComponent} from "./immer-path-ref.js"
export type PathRefProxy<T>={
    __unproxy__:PathRef<T>,
}&{
  [K in keyof T]: PathRefProxy<T[K]>
}&RefProxy<T>
export type ReadonlyPathRefProxy<T>={
    __unproxy__:PathRef<T>,
}&{
  [K in keyof T]: PathRefProxy<T[K]>
}&ReadonlyRefProxy<T>

export function map<T,N>(p:ReadonlyPathRefProxy<T>,m:(o:T)=>N):ReadonlyPathRefProxy<N>{
  return fromR(p.__unproxy__.join(m));
}
export function fromRef<T>(ref:Ref<T>):PathRefProxy<T>{
  const b=c(ref);
  return fromW(b)
}
export function fromVal<T>(val:T):PathRefProxy<T>{
  return fromRef(ref(val))
}
function getBind(target:any, p:any) {
  const r=Reflect.get(target,p);
  if (typeof r==="function") return r.bind(target);
  else return r;
}
function fromW<T>(b:PathRef<T>){
  return new Proxy(b,{
    get(target:PathRef<T>,p:string) {
      if (p==="__unproxy__") {
          return target;
      }
      if (typeof p==="symbol" || p==="then" || p==="toString") {
          return getBind(target,p);
      }
      return fromW(target.join(p as keyof T))
    },
    set(target:PathRef<T>,p:string,v:any){
      if (typeof p==="symbol") {
        Reflect.set(target,p,v);
        return true;
      }
      throw new Error(`Cannot set via ***.${p}. use set( ) instead `)
    },
  }) as any as PathRefProxy<T>
}
function fromR<T>(b:ReadonlyPathRef<T>){
  return new Proxy(b,{
    get(target,p:string) {
      if (p==="__unproxy__") {
          return target;
      }
      if (typeof p==="symbol" || p==="then" || p==="toString") {
          return getBind(target,p);
      }
      return fromW(target.join(p as keyof T))
    },
    set(target:PathRef<T>,p:string,v:any){
      if (typeof p==="symbol") {
        Reflect.set(target,p,v);
        return true;
      }
      throw new Error(`Cannot set via ***.${p}. use set( ) instead `)
    },
  }) as any as ReadonlyPathRefProxy<T>
}
