// @hoge1e3/immer-path-ref/src/proxy.ts
import { RefProxy,ref,Ref,
ReadonlyRefProxy,custom,computed } from "@hoge1e3/ref";
import {produce} from "immer";
import {create as c,Path,
ReadonlyPathRef,applyPath,PathRef,PathComponent} from "./immer-path-ref.js"
export type PathRefProxy<T>={
    __unproxy__:PathRef<T>,
}&{
  [K in keyof T]: PathRefProxy<T[K]>
}&RefProxy<T>
export type ReadonlyPathRefProxy<T>={
    __unproxy__:PathRef<T>,
}&{
  [K in keyof T]: PathRefProxy<T[K]>
}&ReadonlyRefProxy<T>

export function map<T,N>(p:ReadonlyPathRefProxy<T>,m:(o:T)=>N):ReadonlyPathRefProxy<N>{
  return fromR(p.__unproxy__.join(m));
}
export function create<T>(val:T):PathRefProxy<T>{
  const b=c(ref(val))
  return fromW(b)
}
function fromW<T>(b:PathRef<T>){
  return new Proxy(b,{
    get(target:PathRef<T>,p:string) {
      if (p==="__unproxy__") {
          return b;
      }
      if (typeof p==="symbol" || p==="then" || p==="toString") {
          return Reflect.get(b, p);
      }
      return fromW(b.join(p as keyof T))
    }
  }) as any as PathRefProxy<T>
}
function fromR<T>(b:ReadonlyPathRef<T>){
  return new Proxy(b,{
    get(target,p:string) {
      if (p==="__unproxy__") {
          return b;
      }
      if (typeof p==="symbol" || p==="then" || p==="toString") {
          return Reflect.get(b, p);
      }
      return fromW(b.join(p as keyof T))
    }
  }) as any as ReadonlyPathRefProxy<T>
}
