// acepad-widget/acepad-widget.js
import {open} from "acepad-browser";
import {EventHandler} from "acepad-events";
/*global $*/
export function showIframe(url,acepad){
    acepad=acepad||window.sh.$acepad;
    let e=$("<iframe>");
    let title="w";
    if(typeof url==="string"){
        e.attr({src:url});
        title=url.substring(url.length-5);
    }
    if(typeof url.name==="function"){
        title=url.name();
    }
    
    let cont=$("<div>").append(e).attr({title});
    let r=showWidget(cont,acepad);
    if(typeof url.text==="function"){
        //alert(e[0].contentWindow.document);
        open(url,{document:e[0].contentWindow.document});
    }
    
    r.iframe=e[0];
    r.go=(url)=>{
            e.attr({src:url});
        };
    return r;
}
export function showWidget(e,acepad){
    acepad=acepad||window.sh.$acepad;
    const widall=$("<div>").addClass("widget");
    const bar=$("<div>").addClass("widget-bar").appendTo(widall);

    e=$(e);
    e.addClass("widget-container").appendTo(widall);
    let cl=$("<button>").text("x").
    addClass("widget-close").click(close);
    let mi=$("<button>").text("_").
    addClass("widget-minify").click(minify);
    const c=$(acepad.editor.container);
    c.append(widall);
    bar.append(cl);
    bar.append(mi);
    let b=acepad.showMenuButton(e.attr("title")||"w",()=>{
        show();
    });
    function show(){
        widall.show();
        self.fire("show");
    }
    function hide(){
        widall.hide();
        self.fire("show");
    }
    function close(){
        widall.remove();
        b.remove();   
        self.fire("close");
        self.closed=true;
    }
    function minify(){
        widall.hide();
        self.fire("minify");
    }
    const self=EventHandler.attachTo({
        element:e[0],show,
        closeButton:cl,hide,
        close,minify,
    });
    return self;
}
