// acepad-widget/acepad-widget.js
import {open} from "acepad-browser";
import {EventHandler} from "acepad-events";
import {Menu} from "acepad-menu";
import {BoundObject,bind} from "@hoge1e3/dom-bind";
/*global $*/
export function showIframe(url,acepad){
    acepad=acepad||window.sh.$acepad;
    let e=$("<iframe>");
    let title="w";
    if(typeof url==="string"){
        e.attr({src:url});
        title=url.substring(url.length-5);
    }
    if(typeof url.name==="function"){
        title=url.name();
    }
    
    let cont=$("<div>").append(e).attr({title});
    let r=showWidget(cont,acepad);
    if(typeof url.text==="function"){
        //alert(e[0].contentWindow.document);
        open(url,{document:e[0].contentWindow.document});
    }
    
    r.iframe=e[0];
    r.go=(url)=>{
            e.attr({src:url});
        };
    return r;
}
export function showWidget(e,opt={}){
    e=$(e);

    let {acepad,buttons,noMenu}=opt;
    buttons=buttons||[
        {command:"close",caption:"x"},
        {command:"minify",caption:"_"},
    ];
    acepad=acepad||window.sh.$acepad;
    let menu=Menu(acepad.getMainEditor());
    const widall=$("<div>").addClass("widget");
    const bar=$("<div>").addClass("widget-bar").appendTo(widall);

    e.addClass("widget-container").appendTo(widall);
    const c=$(acepad.getMainEditor().container);
    c.append(widall);
    for(let k of buttons)btn(k);
    function btn({command,caption}){
        let b=$("<button>").text(caption);
        if(typeof command==="string"){
            let c=command;
            command=()=>self[c]();
        }
        b.addClass("widget-button").click(command);
        bar.append(b);
    }
    const title=e.attr("title")||"w";
    let menub=!noMenu&&menu.add(title, show);
    function show(){
        widall.show();
        self.fire("show");
    }
    function hide(){
        self.minify();
    }
    function close(){
        widall.remove();
        if(menub)menub.remove();   
        self.fire("close");
        self.closed=true;
    }
    function minify(){
        widall.hide();
        self.fire("minify");
    }
    const self=EventHandler.attachTo({
        show,
        element:e[0],
        //closeButton:cl,
        hide,
        close,minify,
    });
    widall[0][BoundObject]=self;
    return self;
}
export function minifyAll(){
    $(".widget").each(function (){
       this[BoundObject].minify(); 
    });
}
