// @acepad/with-fs/acepad-with-fs.js
// acepad-with-fs
import {open} from "@acepad/browser";
import {trial,debugSession} from "@acepad/debug";
import {showWidget} from "@acepad/widget";
import {initVConsole,showVConsole} from "show-vconsole";
import {run} from "@acepad/find";
import unload from "@hoge1e3/unload";
import {Menu} from "@acepad/menu";
import {openFile,createSessionListWithRecents} from "@acepad/files";
import {init} from "acepad";
import * as fkeys from "@acepad/function-keys";
import {loadScriptTag} from "@hoge1e3/loadScript";
import {wrap} from "@hoge1e3/splashscreen";
import {gen,fix} from "@acepad/genai";
/*global $*/
async function loadScripts(scripts) {
    return Promise.all(scripts.
    map(s=>`https://unpkg.com/${s}`).
    map(async (u)=>{
        try{
            //console.log("Loading ",u);
            await loadScriptTag(u);
            //console.log("Loaded ",u);
        }catch(e){
            alert("load fail"+e.message);
        }
    }));
}
export async function main(){
    const sh=this;
    const dirn=sh.resolve(
        import.meta.url
    ).up();
    if (sh.$home){
        const dir_nodem=sh.resolve(sh.$home).rel("node_modules/");
        setPath(sh,dir_nodem);    
    }
    await loadScripts([
"jquery@1.12.1/dist/jquery.min.js",
"vconsole@latest/dist/vconsole.min.js",
"ace-builds@1.39.0/src-noconflict/ace.js",
]);
    await loadScripts([
"ace-builds@1.39.0/src-noconflict/ext-language_tools.js",
    ]);
    console.log("Render HTML");
    await open(dirn.rel("ace.html"));
    const {setForce}=unload({
        oncancel(){
           setTimeout(()=> 
          document.documentElement.requestFullscreen().then(
              ()=>document.exitFullscreen())
        ,100);  
        }
    });
    const acepad=window.acepad=await init();   
    console.log("Render Done");
    sh.$acepad=window.acepad;
    sh.setenv("ROOT_SHELL",sh);
    sh.setenv("ACEPAD_INSTANCE",acepad);
    
    //let {openFile}=await sh.jsm("acepad-files");
    acepad.attachCommands({
        F1:"changeSession",
        F5:"runMenu",
        F7:"dirup",
        F11:"showConsole",
        F12:"reloadPage",
        "ctrl-g":"toggleGutter",
        "ctrl-9":"toggleFullScreen",
        "ctrl-f":"findFiles",
        
        reloadPage(){location.reload();},
        findFiles(){run(sh);},
        dirup:()=>{
            let si=acepad.sessionInfo(
                acepad.getMainEditor().session);
            if(si.file){
                openFile(sh, si.file.up());
            }
        },
        showConsole:()=>{
            showVConsole();
        },
        toggleGutter:()=>{
            //autogutt disable this
            const editor=acepad.getCurrentEditor();
            //console.log("gutt",editor.renderer.getShowGutter());
            editor.renderer.setShowGutter(
                !editor.renderer.getShowGutter());
        },
        toggleFullScreen() {
          if (!document.fullscreenElement) {
            document.documentElement.requestFullscreen();
          } else if (document.exitFullscreen) {
            document.exitFullscreen();
          }
        },
        changeSession(){
            acepad.changeSession(rs); 
        },
        runMenu(){
            sh.runMenu();
        },
        genAi(){
            gen(sh);
        },
        fixAi(){
            fix(sh);
        },
    });
    setTimeout(initVConsole,1000);
    initMenuButtons(acepad);
    sh.debugSession=()=>debugSession(sh);
    autogutt(acepad);
    fkeys.show(acepad);
    const rs=createSessionListWithRecents(sh);
    acepad.changeSession(rs);
    sh.keysel(1);
    sh.quota();
    hackTimeouts();
    return acepad;
}
function setPath(sh,nodem){
    sh.addPath(nodem.rel(".bin/").path());
    //sh.addPath(sh.resolve("bin/").path());
}
let prevs;
export function initMenuButtons(acepad){
    let {events,sessionInfo,changeSession}=acepad;
    let menu=Menu(acepad.getMainEditor());
    return;
    events.on("createSession",({name,session})=>{
        //console.log(name);
        let m;
        let si=sessionInfo(session);
        si.on("infochanged",(d)=>{
            m.text(d.name);
        });
        m=menu.add(name,()=>{
            changeSession(session);
        }); 
        si.on("remove",()=>{
            m.remove();
        });
        m.go2nd();
        menu.sort();
        si.menuButton=m;
    });
    events.on("changeSession",({oldSession,session})=>{
        let si=sessionInfo(session);
        let osi=sessionInfo(oldSession);
        if(!si.menuButton||!osi.menuButton)return ;
        $(osi.menuButton.element).removeClass("selected");
        $(si.menuButton.element).addClass("selected");
        si.menuButton.setOrder(0);
        osi.menuButton.setOrder(0);
        
        menu.sort();
    });
    menu.add({
       /* cons(){
            showVConsole();
        },
        ses(){
            acepad.changeSession("*sessions*");
        },
        gutt(){
            const editor=acepad.getCurrentEditor();
            editor.renderer.setShowGutter(
                !editor.renderer.getShowGutter());
        },*/
    });
}
function autogutt(acepad){
    //!run
    const renderer=acepad.getMainEditor().renderer;
    setInterval( ()=>{
        const g=renderer.getShowGutter();
         if (g&&renderer.getScrollLeft()>50) {
            renderer.setShowGutter(false);
        } else  if (!g&&renderer.getScrollLeft()<1) {
            renderer.setShowGutter(true);
        }
   
    },100);
    
}
function hackTimeouts(){
    for(let k of ["setTimeout","setInterval",
    "clearTimeout","clearInterval",]){
        globalThis[k]=globalThis[k].bind(globalThis);
    }
}
/*
function runMenu(sh){
    const acepad=sh.$acepad;
    const si=acepad.sessionInfo(
        acepad.getMainEditor().session);
    if(si.file){
        const nsh=sh.clone();
        nsh.cd(si.file.isDir()?
            si.file:si.file.up());
        let cmd=nsh.hasExecutableCommand(si.file);
        if(!cmd){
            for(let p=si.file.up();p;p=p.up()){
                const f=(p.rel(".runMenu"));
                if(f.exists()){
                    nsh.cd(p);
                    cmd=`./.runMenu ${si.file.path()}`;
                    break;
                }
            }
        }
        if(!cmd){
            const d=nsh["npm-find"](si.file);
            if(d){
                cmd="npm-test";
            }
        }
        if(cmd){
            //console.log("pwd",nsh.pwd());
            const ds=debugSession(nsh,cmd);
            //acepad.changeSession(ds);//
            return ;
        }
    }
    alert("No run command for "+si.name);
}*/
