// @acepad/mcp/mcp.js
#!run
// @acepad/mcp

import {t} from "@hoge1e3/dom";
import {show} from "@acepad/widget";
import * as patch from "@hoge1e3/patch";

function read(f){
  if (f.ext()===".json") {
    try {
      return f.obj();
    } catch(e){
      
    }
  }
  return patch.stripEmptyLines(f.text());
}
function d2j(d){
  const r={};
  for(let f of d.recursive({excludes:["test/"]})){
    if(f.endsWith(".js.map")) continue;
    r[f.relPath(d)] = read(f);    
  }
  return r;
}

export async function main(){
  const sh=this;
  const d=sh.cwd();
  
  const w=show({overlay:true});

  let userPrompt="";

  function buildPrompt(){
    return `
${userPrompt}

You are generating a git unified diff patch.

Strict rules:
- Output must be a valid unified diff.
- Do not include markdown.
- Do not wrap in code fences.
- Do not include explanation.
- The output must be directly applicable using \`git apply\`.
- Use the exact format:
  diff --git a/<path> b/<path>
  --- a/<path>
  +++ b/<path>
  @@ -old_line,old_count +new_line,new_count @@
- If the changes are needed on multiple files, repeat the format above.
- These entire patches should be shown with single code snippet.

~~~json
${JSON.stringify(d2j(d))}
~~~
  `;
    }
  
    const a=(f,c)=>t.a({href:"javascript:;",onclick:f},c);
  
    w.print(
      t.div({id:"mcp-panel",style:"height:100%; overflow: scroll;"},
        t.div("Custom Prompt"),
        //prompt: fit textarea size into widget size
        t.div(t.textarea({
          rows: 8,
          cols: 100,
          style:"width:100%; height:30%; box-sizing:border-box;",
          oninput(e){
            userPrompt=e.target.value;
          }
        },userPrompt)),
        t.ol(
          t.li(a(cp,"Copy prompt")),
          t.li(
            t.a({
              target:"ai", 
              href:process.env.genai||"https://chatgpt.com/"
            },"Open GenAI"),
            " and paste prompt."
          ),
          t.li(a(paste,"Paste result")),
        ),
      )
    );
  
    function cp(){
      const pr=buildPrompt();
      navigator.clipboard.writeText(pr);
    }
  
  // Render a simple colored diff view for a hunk
  function renderDiff(hunks, patchText, filePath) {
    const container = t.div({style:"font-family:monospace; font-size:12px; white-space:pre; overflow-x:auto;"});
    const patchLines = patchText.split("\n");
    let inFile = false;
    let inHunk = false;

    for (const line of patchLines) {
      if (line.startsWith("+++ b/" + filePath)) {
        inFile = true;
        continue;
      }
      if (!inFile) continue;
      if (line.startsWith("diff --git ")) break;

      if (line.startsWith("@@ ")) {
        inHunk = true;
        container.appendChild(t.div({style:"color:#888; background:#f0f0f0;"}, line));
        continue;
      }
      if (!inHunk) continue;

      let style = "";
      if (line.startsWith("+")) style = "color:#006400; background:#e6ffe6;";
      else if (line.startsWith("-")) style = "color:#8b0000; background:#ffe6e6;";
      else style = "color:#333;";
      container.appendChild(t.div({style}, line));
    }
    return container;
  }

  async function paste(){
      const s=await navigator.clipboard.readText();
      // s contains patch format
      const files = patch.parse(s);

      if (files.length === 0) {
        w.print(t.div({id:"mcp-panel", style:"padding:8px;"},
          t.div("No valid unified diff found in clipboard.")
        ));
        return;
      }

      const panel = t.div({id:"mcp-panel", style:"height:100%; overflow:scroll; padding:8px;"});
      panel.appendChild(t.div({style:"font-weight:bold; margin-bottom:8px;"}, "Result — Review & Accept"));

      for (const filePatch of files) {
        if (!filePatch.path) continue;

        const fileSection = t.div({style:"border:1px solid #ccc; margin-bottom:12px; border-radius:4px; overflow:hidden;"});

        // Header bar with filename and Accept button
        const header = t.div({style:"display:flex; align-items:center; justify-content:space-between; background:#f5f5f5; padding:4px 8px; border-bottom:1px solid #ccc;"});
        header.appendChild(t.span({style:"font-family:monospace; font-weight:bold;"}, filePatch.path));

        const acceptBtn = t.button({
          style:"margin-left:8px; padding:2px 10px; background:#4caf50; color:#fff; border:none; border-radius:3px; cursor:pointer;",
          onclick: async () => {
            try {
              const f = d.rel(filePatch.path);
              const original = f.exists() ? patch.stripEmptyLines(f.text()) : "";
              const newContent = patch.apply(original, filePatch.hunks, s, filePatch.path);
              console.log(f.path(),newContent);
              f.text(newContent);
              acceptBtn.textContent = "Accepted ✓";
              acceptBtn.disabled = true;
              acceptBtn.style.background = "#888";
            } catch(e) {
              acceptBtn.textContent = "Error!";
              acceptBtn.title = String(e);
              acceptBtn.style.background = "#c00";
            }
          }
        }, "Accept");

        header.appendChild(acceptBtn);
        fileSection.appendChild(header);

        // Diff view
        fileSection.appendChild(renderDiff(filePatch.hunks, s, filePatch.path));
        panel.appendChild(fileSection);
      }

      // Accept All button
      const allBtn = t.button({
        style:"margin-top:4px; padding:4px 16px; background:#2196f3; color:#fff; border:none; border-radius:3px; cursor:pointer;",
        onclick: () => {
          document.querySelector("#mcp-panel").
          querySelectorAll("button:not([disabled])").forEach(b => {
            if (b !== allBtn) b.click();
          });
        }
      }, "Accept All");
      panel.appendChild(allBtn);

      w.print(panel);
    }
  }
