// @acepad/mcp/mcp.js
#!run
// @acepad/mcp

import {t} from "@hoge1e3/dom";
import {show} from "@acepad/widget";

function d2j(d){
  const r={};
  for(let f of d.recursive({excludes:["test/"]})){
    if(f.endsWith(".js.map")) continue;
    r[f.relPath(d)] = f.ext()===".json" ? f.obj() : f.text();    
  }
  return r;
}

export async function main(){
  const sh=this;
  const d=sh.cwd();
  
  const w=show({overlay:true});

  let userPrompt="";

  function buildPrompt(){
    return `
${userPrompt}

You must answer in pure JSON.
Do not output explanation.
Keys are file paths.
Values are full new content of file(double-quotes should be escaped) .

You are generating a machine-readable patch.

Rules:
- Output exactly one JSON object.
- Do not output markdown.
- Do not wrap in code fences.
- Do not output explanation.
- The first character must be {.
- The last character must be }.
- The result must be valid JSON.parse input.
- Escape all double quotes inside file contents properly.

Format:
{
  "path/to/file": "import * as fs from \\"fs\\";"
}

~~~json
${JSON.stringify(d2j(d))}
~~~
  `;
    }
  
    const a=(f,c)=>t.a({href:"javascript:;",onclick:f},c);
  
    w.print(
      t.div({style:"height:100%; overflow: scroll;"},
        t.div("Custom Prompt"),
        //prompt: fit textarea size into widget size
        t.div(t.textarea({
          rows: 8,
          cols: 100,
          oninput(e){
            userPrompt=e.target.value;
          }
        },userPrompt)),
        t.ol(
          t.li(a(cp,"Copy prompt")),
          t.li(
            t.a({
              target:"ai", 
              href:process.env.genai||"https://chatgpt.com/"
            },"Open GenAI"),
            " and paste prompt."
          ),
          t.li(a(ps,"Paste result")),
        ),
      )
    );
  
    function cp(){
      const pr=buildPrompt();
      navigator.clipboard.writeText(pr);
    }
  
    async function ps(){
      const s=await navigator.clipboard.readText();
      let o;
      try{
        o=JSON.parse(s);
      }catch(e){
        alert("Invalid JSON");
        return;
      }
    
      const keys=Object.keys(o);
      let k,f;
    
      const str=(s)=>typeof s==="string"?s:JSON.stringify(s,null,4);
    
      function showResult(_k){
        k=_k;
        f=k && d.rel(k);
      
        w.print(
          t.div({style:"height:100%; overflow: scroll;"},
            t.div("Result"),
            ...keys.map(k=>
              t.div(a(()=>{showResult(k)},k))
            ),
            ...(f ? [
              t.div(t.button({
                onclick(){
                  f.text(str(o[k]));
                  keys.splice(keys.indexOf(k),1);
                  showResult();
                },
              },`Accept ${k}`)),
              t.div(t.textarea({
                rows: 20,
                cols: 100,
              },str(o[k])))
            ] : []),
          )
        );
      }
    
      showResult();
    }
  }
