// @acepad/keypad/js/modifier.js
import {sync} from "@hoge1e3/key-stream";

export var modifierState={},mLock={
    sym:1
};
export var modifierKeys={"shift":1,"ctrl":1,"edit":1,"sym":1,"select":1};
export function currentModifierState(){
    return modifierState;
}
export function modifierStateToInt(s){
    s=s||modifierState;
    return ssToInt(s);
}
export function modifierStateToKeyboardEvent(s){
  s=s||modifierState;
  return {
    shiftKey:s.shift,
    ctrlKey:s.ctrl,
    
  }; 
}
export function syncAsKeyboardEvent(s){
  sync(modifierStateToKeyboardEvent(s));
}
export function setModifier(k,v){
    switch(v){
        case 0:
            modifierState[k]=0;
            mLock[k]=0;
            break;
        case 1:
            modifierState[k]=1;
            mLock[k]=0;
            break;
        case 2:
            modifierState[k]=1;
            mLock[k]=1;
            break;
    }
}
export function getModifier(k){
    return modifierState[k];
}
export function modifierLocked(k){
    return mLock[k];
}
export function clearUnlockedModifiers(){
    for(let k in modifierState){
        if(modifierLocked(k))continue;
        modifierState[k]=0;
    }
}
export function ssToInt(e){
    return 0 | 
    (e.ctrl ? 1 : 0) | 
    (e.alt ? 2 : 0) | 
    (e.shift ? 4 : 0) | 
    (e.meta ? 8 : 0);
}
//------keypad 
// moved to keypad.js

export function modifierButtonPressed(d){
        let mod=d.modifier;

    let lock=d.lock||"double";
    if(!modifierState[mod]){
        modifierState[mod]=1;
        if(lock=="single"){
            mLock[mod]=1;
        }
    }else if(lock!="none"&&
    !mLock[mod]){
        mLock[mod]=1;
    }else{
        modifierState[mod]=
        mLock[mod]=0;
    }
    if(d["modifier-off"]){
        modifierState[d["modifier-off"]]=0;
    }
    return;
}
export function addModifierStyle(){
  const styleEl = document.createElement("style");
  styleEl.setAttribute("id","sutairu");
  // Append <style> element to <head>
  document.head.appendChild(styleEl);
  // Grab style element's sheet
  const styleSheet = styleEl.sheet;
  //styleSheet.insertRule("button{color:blue;}");
  for(let k in modifierKeys){
    styleSheet.insertRule(`#keypad.${k} .mask.no-${k}{
        display: none;
    }`);
    styleSheet.insertRule(`#keypad:not(.${k}) .mask.${k}{
        display: none;
    }`);
  }
}
