// @acepad/os/index.js
/*global globalThis*/
import {loadedModules} from "petit-node"
export {importModule, FS, urlToPath, require} from "petit-node";
let _convertRequired;
export let hasMessageInStack;
function convertRequired() {
  if (typeof _convertRequired==="boolean")return _convertRequired;
  const mesg="this is test message";
  try {
    throw new Error(mesg);
  } catch(e) {
    if (e.stack.includes(mesg)) {
      hasMessageInStack=true;
      _convertRequired=false;
    } else {
      hasMessageInStack=false;
      _convertRequired=true;
    }
    
  }
  return _convertRequired;
}
convertRequired();
export const convertStack=(e, force=false)=>{
  if (!force && !convertRequired()) return e;
  if(!e) return e;
  if (typeof e.stack!=="string") return e;
  for (let m of loadedModules()) {
    if (!m.url)continue;
    let i=e.stack.indexOf(m.url);
    if (i<0) continue;
    e.stack=e.stack.substring(0,i)+
    (m.entry?"file://":"")+m.path+e.stack.substring(i+m.url.length);
  }
  e.stack=e.message+"\n"+e.stack;
  return e;
};    

//export const resolveModule=o.resolveModule;
