// @acepad/store-file/acepad-store-file.js
// acepad-store-file
import FS from "@hoge1e3/fs";
import * as s from "acepad-store";

const tmpath="/tmp/";
export function validate(data,noid){
    if((!noid &&!data.__id__) || !data.url){
        console.log("data",data);
        throw new Error("no __id__ or url "+noid);
    }
    console.log("datavalid",data);
    return data;
}
export async function put(src){
    const data=file2data(src);
    const id=await s.put(data);
    data.__id__=id;
    validate(data);
    return id;
}
export async function get(id,dst){
    let data=await s.get(id);
    data2file(validate(data),dst);
}
export function data2file(data,dst){
    dst.dataURL(validate(data).url);
}
export function file2data(src){
    return validate({url:src.dataURL()},true);
}
export function zip2data(src){
    return validate(file2data(src),true);
}
export function data2zip(data){
    validate(data);
    let ram=FS.get(tmpath);    
    let zip=ram.rel(data.__id__+".zip");
    data2file(data,zip);
    return zip;
}
export async function data2dir(data){
    validate(data);
    let zip=await data2zip(data);
    let ext=zip.sibling(data.__id__+"/");
    await FS.zip.unzip(zip,ext,{overwrite:true});
    return ext;
}
export async function dir2data(dir,opt={}){
    let ram=FS.get(tmpath);    
    let n=Math.floor(Math.random()*100000);
    let zip=ram.rel(n+".zip");
    await FS.zip.zip(dir,zip,opt);
    let data=zip2data(zip);
    return validate(data,true);
}
