// @hoge1e3/ref/js/ref.js
// @hoge1e3/ref
import { EventHandler } from "@hoge1e3/events";
export const V = Symbol.for("ref_get");
export const E = Symbol.for("ref_change");
export const F = Symbol.for("ref_fire");
const is_hoge1e3 = Symbol("is_hoge1e3");
export function ref(val) {
    const r = custom(() => val, (n) => val = n);
    return r;
}
export function get(r) {
    return r[V];
}
export function set(r, v) {
    r[V] = v;
}
export function change(r, h) {
    r[E](h);
}
export function computed(...args) {
    //if(typeof a[a.length-2]!="function"){
    //args.push(()=>throw new Error())  
    //}
    let f = args.pop(); //(...values:RefValues<RS>)=>T;
    //if(args[])
    let val;
    let ti;
    function h() {
        const prev = val;
        val = f(...args.map((r) => get(r)));
        if (prev !== val) {
            res.fire("change", { prev, val });
        }
        ti = null;
    }
    const req = () => {
        if (ti != null)
            return;
        ti = setTimeout(h, 0);
    };
    for (let r of args) {
        r.on("change", req);
    }
    const res = EventHandler.attachTo({
        get() {
            return val;
        },
        get value() {
            return res.get();
        },
        get [V]() {
            return res.get();
        },
        [E](h) {
            return res.on("change", h);
        },
        [F](event, payload) {
            res.fire(event, payload);
        }
    }); // as ReadonlyRef<T>;
    h();
    return res;
}
export function custom(g, s) {
    const res = EventHandler.attachTo({
        [is_hoge1e3]: true,
        get: g,
        set(n) {
            const val = g();
            if (val === n)
                return;
            const prev = val;
            s(n);
            res.fire("change", {
                prev, val: n,
            });
        },
        get value() {
            return g();
        },
        set value(v) {
            this.set(v);
        },
        get [V]() {
            return g();
        },
        set [V](v) {
            this.set(v);
        },
        [E](h) {
            return res.on("change", h);
        },
        [F](event, payload) {
            res.fire(event, payload);
        }
    });
    return res;
}
