// @acepad/tsc/src/runts.js
#!run
import ts from "typescript";
import * as path from "path";
// implement readTsconfig and call from getJS
// implement readTsconfig and call from getJS
function readTsconfig(srcfile) {
    const configPath = ts.findConfigFile(path.dirname(srcfile), ts.sys.fileExists, "tsconfig.json");
    console.log("configPath", configPath);
    if (!configPath) {
        return undefined;
    }
    const configFile = ts.readConfigFile(configPath, ts.sys.readFile);
    const config = ts.parseJsonConfigFileContent(configFile.config, ts.sys, path.dirname(configPath));
    console.log("config", config);
    return config;
}
export function getJS(tsfile) {
    if (tsfile.endsWith(".js"))
        return tsfile;
    const config = readTsconfig(tsfile);
    if (!config) {
        return tsfile.replace(/\.ts$/, ".js");
    }
    const { outDir, rootDir } = config.options;
    if (!outDir) {
        return tsfile.replace(/\.ts$/, ".js");
    }
    const root = rootDir || getCommonDirectory(config.fileNames);
    const rel = path.relative(root, tsfile);
    console.log(tsfile, root, rootDir, outDir, rel);
    return path.join(outDir, rel).replace(/\.ts$/, ".js");
}
export function getTS(jsfile) {
    if (jsfile.endsWith(".ts"))
        return jsfile;
    const config = readTsconfig(jsfile);
    if (!config) {
        return jsfile.replace(/\.js$/, ".ts");
    }
    const { outDir, rootDir } = config.options;
    if (!outDir) {
        return jsfile.replace(/\.js$/, ".ts");
    }
    const root = rootDir || getCommonDirectory(config.fileNames);
    const rel = path.relative(outDir, jsfile);
    return path.join(root, rel).replace(/\.js$/, ".ts");
}
function getCommonDirectory(paths) {
    if (paths.length == 1) {
        return path.dirname(paths[0]);
    }
    const [first, ...rest] = paths.map(p => path.resolve(p));
    let common = first;
    for (const p of rest) {
        common = commonPath(common, p);
    }
    return common;
}
function commonPath(a, b) {
    const aParts = a.split(path.sep);
    const bParts = b.split(path.sep);
    const len = Math.min(aParts.length, bParts.length);
    let i = 0;
    while (i < len && aParts[i] === bParts[i]) {
        i++;
    }
    return aParts.slice(0, i).join(path.sep) || path.sep;
}
export async function main(path, ...a) {
    const jsf = getJS(this.resolve(path).path());
    console.log("js file:", path, jsf);
    return await this.run(jsf, ...a);
}
