// @acepad/tsc/src/watchworker.ts
#!run 
import {retry} from "@acepad/retry";
import ts from "typescript";
const formatHost: ts.FormatDiagnosticsHost = {
  getCanonicalFileName: path => path,
  getCurrentDirectory: ts.sys.getCurrentDirectory,
  getNewLine: () => ts.sys.newLine
};
let x:number=50;
let y:string="23";
//_this is comment
export async function main(dir:string,ch?:MessagePort) {
  await retry(()=>
  process.chdir(dir));
    //this.echo(process.cwd());
  const configPath = ts.findConfigFile(
    /*searchPath*/ "./",
    ts.sys.fileExists,
    "tsconfig.json"
  );//q
  if (!configPath) {
    throw new Error("Could not find a valid 'tsconfig.json'.");
  }
  // TypeScript can use several different program creation "strategies":
  //  * ts.createEmitAndSemanticDiagnosticsBuilderProgram,
  //  * ts.createSemanticDiagnosticsBuilderProgram
  //  * ts.createAbstractBuilder
  // The first two produce "builder programs". These use an incremental strategy
  // to only re-check and emit files whose contents may have changed, or whose
  // dependencies may have changes which may impact change the result of prior
  // type-check and emit.
  // The last uses an ordinary program which does a full type check after every
  // change.
  // Between `createEmitAndSemanticDiagnosticsBuilderProgram` and
  // `createSemanticDiagnosticsBuilderProgram`, the only difference is emit.
  // For pure type-checking scenarios, or when another tool/process handles emit,
  // using `createSemanticDiagnosticsBuilderProgram` may be more desirable.
    console.log("configPath",configPath);
  const createProgram = ts.createSemanticDiagnosticsBuilderProgram.bind(ts);

  // Note that there is another overload for `createWatchCompilerHost` that takes
  // a set of root files.
  const host = ts.createWatchCompilerHost(
    configPath,
    {},
    ts.sys,
    createProgram,
    reportDiagnostic,
    reportWatchStatusChanged
  );
  // You can technically override any given hook on the host, though you probably
  // don't need to.
  // Note that we're assuming `origCreateProgram` and `origPostProgramCreate`
  // doesn't use `this` at all.
  let diags=[] as any[];
  const origCreateProgram = host.createProgram.bind(host);
  //error: A spread argument must either have a tuple type or be passed to a rest parameter.
  host.createProgram = (...a:Parameters<typeof origCreateProgram>
  /*rootNames: ReadonlyArray<string>, 
  options, host, oldProgram*/) => {
    diags=[];
    console.log("** We're about to create the program! **");
    return origCreateProgram(...a
    /*rootNames, options, host, oldProgram*/);
  };
  const origPostProgramCreate = host.afterProgramCreate?.bind(host);

  host.afterProgramCreate = program => {
    console.log("** We finished making the program! **");
    if(origPostProgramCreate)origPostProgramCreate!(program);
    console.log("diags",diags);
    ch?.postMessage(diags);
  };

  // `createWatchProgram` creates an initial program, watches files, and updates
  // the program over time.
 // this.echo("c3");
  ts.createWatchProgram(host);

  
  function reportDiagnostic(diagnostic: ts.Diagnostic) {
   const d={
        start:diagnostic.start,
        length:diagnostic.length,
        file:diagnostic.file?.fileName,
        messageText:diagnostic.messageText,
        
      };
      diags.push(d);
     // ch?.postMessage(toAnnotations(diagnostic));
    console.error("Error", 
    diagnostic.code, ":", 
    ts.flattenDiagnosticMessageText( 
        diagnostic.messageText, formatHost.getNewLine()));
  }
    
}

/**
 * Prints a diagnostic every time the watch status changes.
 * This is mainly for messages like "Starting compilation" or "Compilation completed".
 */
function reportWatchStatusChanged(diagnostic: ts.Diagnostic) {
  console.info(ts.formatDiagnostic(diagnostic, formatHost));
}

function toAnnotations(diagnostics: ts.Diagnostic[]){

    return diagnostics.map(diag => {
      if(!diag.file||!diag.start)return {};
        const { line, character } = diag.file.getLineAndCharacterOfPosition(diag.start);
        return {
            row: line,
            column: character,
            text: ts.flattenDiagnosticMessageText(diag.messageText, "\n"),
            type: "error"
        };
    });
}


//watchMain();
