// @acepad/tsc/src/watchworker.js
#!run 
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { retry } from "@acepad/retry";
import ts from "typescript";
//import * as fs from "fs";
const formatHost = {
    getCanonicalFileName: path => path,
    getCurrentDirectory: ts.sys.getCurrentDirectory,
    getNewLine: () => ts.sys.newLine
};
let x = 50;
let y = "23";
//_this is comment
export function main(dir, ch) {
    return __awaiter(this, void 0, void 0, function* () {
        var _a;
        yield retry(() => process.chdir(dir));
        //this.echo(process.cwd());
        const configPath = ts.findConfigFile(
        /*searchPath*/ "./", ts.sys.fileExists, "tsconfig.json"); //q
        if (!configPath) {
            throw new Error("Could not find a valid 'tsconfig.json'.");
        }
        // TypeScript can use several different program creation "strategies":
        //  * ts.createEmitAndSemanticDiagnosticsBuilderProgram,
        //  * ts.createSemanticDiagnosticsBuilderProgram
        //  * ts.createAbstractBuilder
        // The first two produce "builder programs". These use an incremental strategy
        // to only re-check and emit files whose contents may have changed, or whose
        // dependencies may have changes which may impact change the result of prior
        // type-check and emit.
        // The last uses an ordinary program which does a full type check after every
        // change.
        // Between `createEmitAndSemanticDiagnosticsBuilderProgram` and
        // `createSemanticDiagnosticsBuilderProgram`, the only difference is emit.
        // For pure type-checking scenarios, or when another tool/process handles emit,
        // using `createSemanticDiagnosticsBuilderProgram` may be more desirable.
        console.log("configPath", configPath);
        const createProgram = ts.createSemanticDiagnosticsBuilderProgram.bind(ts);
        // Note that there is another overload for `createWatchCompilerHost` that takes
        // a set of root files.
        const host = ts.createWatchCompilerHost(configPath, {}, ts.sys, createProgram, reportDiagnostic, reportWatchStatusChanged);
        // You can technically override any given hook on the host, though you probably
        // don't need to.
        // Note that we're assuming `origCreateProgram` and `origPostProgramCreate`
        // doesn't use `this` at all.
        let diags = [];
        const origCreateProgram = host.createProgram.bind(host);
        //error: A spread argument must either have a tuple type or be passed to a rest parameter.
        host.createProgram = (...a
        /*rootNames: ReadonlyArray<string>,
        options, host, oldProgram*/ ) => {
            diags = [];
            console.log("** We're about to create the program! **");
            return origCreateProgram(...a
            /*rootNames, options, host, oldProgram*/ );
        };
        const origPostProgramCreate = (_a = host.afterProgramCreate) === null || _a === void 0 ? void 0 : _a.bind(host);
        host.afterProgramCreate = program => {
            console.log("** We finished making the program! **");
            if (origPostProgramCreate)
                origPostProgramCreate(program);
            console.log("diags", diags);
            ch === null || ch === void 0 ? void 0 : ch.postMessage(toAnnotations(diags));
        };
        // `createWatchProgram` creates an initial program, watches files, and updates
        // the program over time.
        // this.echo("c3");
        ts.createWatchProgram(host);
        function ms(messageText) {
            if (typeof messageText === "string") {
                return messageText;
            }
            return messageText.messageText + "+";
        }
        function reportDiagnostic(diagnostic) {
            /*const d={
                 start:diagnostic.start,
                 length:diagnostic.length,
                 file:diagnostic.file,//?.fileName,
                 messageText:ms(diagnostic.messageText),
                 
               };*/
            diags.push(diagnostic);
            // ch?.postMessage(toAnnotations(diagnostic));
            console.error("Error", diagnostic.code, ":", ts.flattenDiagnosticMessageText(diagnostic.messageText, formatHost.getNewLine()));
        }
    });
}
/**
 * Prints a diagnostic every time the watch status changes.
 * This is mainly for messages like "Starting compilation" or "Compilation completed".
 */
function reportWatchStatusChanged(diagnostic) {
    console.info(ts.formatDiagnostic(diagnostic, formatHost));
}
function toAnnotations(diagnostics) {
    return diagnostics.filter(diag => {
        if (!diag.file || diag.start === undefined)
            return false;
        return true;
    }).map(diag => {
        const { line, character } = ts.getLineAndCharacterOfPosition(diag.file, diag.start);
        return {
            //start: diag.start,
            file: diag.file.fileName,
            row: line,
            column: character,
            text: ts.flattenDiagnosticMessageText(diag.messageText, "\n"),
            type: "error"
        };
    });
}
/*function rowcol(file:string,
idx:number) {
  let lines=fs.readFileSync(file,{encoding:"utf8"}).
  split("\n");
  let row=0;
  for (let line of lines) {
    line+="\n";
    idx-=line.length;
    if (idx<=0) {
      return {row, column:line.length+idx};
    }
    row++;
  }
  return {row,column:0};
}
*/
//watchMain();
