// @hoge1e3/rpc/src/proxy.ts
import { Client } from "./client.js";
import { Server } from "./server.js";
import { isMessegable, Messagable } from "./types.js";

type Methods={[key:string]:Function};
export function server(
    target: Messagable, 
    channel:string,
    allowOrigin:string[],
    methods:Methods,
):Server;
export function server(
    channel:string,
    allowOrigin:string[],
    methods:Methods,
):Server;
export function server(...a:any[]):Server {
    const target:Messagable=(isMessegable(a[0])?
        a.shift():
        globalThis as unknown as Messagable
    );
    const channel:string=a.shift();
    const allowOrigin:string[]=a.shift();
    const methods:Methods=a.shift();
    const s=new Server(target, channel, allowOrigin);
    for (let k in methods) {
        s.serv(k,(params)=>methods[k](...params.args));
    }
    return s;
}
export function client(     
    target: Messagable, 
    channel="default",
    origin=undefined as (string|undefined),
    manualProbe=false, 
) {
    const c=new Client(target, channel, origin, manualProbe);
    return new Proxy(c,{
        get(target,p) {
            if (p==="__unproxy__") {
                return c;
            }
            if (typeof p==="symbol" || p==="then" || p==="toString") {
                return Reflect.get(target, p);
            }
            //const t=args.map((a:any)=>transfers.get(a))
            return (...args:any[])=>{
              const t=args.reduce(
                (tr,a:any)=>
                [...t,...transfers.get(a)??[]])
              return c.run(p,{args});
            }
        }
    })
}
const transfers=new WeakMap<any,any[]>();
export function transfer<T>(o:T,t?:any[]):T{
  if(!t)t=[o];
  transfers.set(o,t)
  return o;
}

type PeerParamBase={
    channel: string, 
    service:{[key:string]: Function},      
};
type PeerParam=WorkerPeerParam|XOriginPeerParam;
type WorkerPeerParam=PeerParamBase&{
    target: Worker,
};
function isWorkerPeerParam(p:PeerParam):p is WorkerPeerParam {
    return p.target instanceof Worker;
}
type XOriginPeerParam=PeerParamBase&{
    target: Messagable,
    targetOrigin: string,
    allowOrigin?: string[],
};
export function peer(param:PeerParam) {
    const [targetOrigin, allowOrigin]=isWorkerPeerParam(param)?
    [undefined, [] as string[]]:
    [param.targetOrigin, param.allowOrigin?? [param.targetOrigin]];
    const {target, service, channel}=param;
    const cli=client(target, channel, targetOrigin, false);
    const srv=server(globalThis as Messagable, channel, allowOrigin, service);
    (cli as any).__unproxy__.addEventListener("dispose", ()=>srv.dispose());
    return cli;
}
