// @acepad/todo/js/todo.js
// @acepad/todo
import { id as idg } from "@hoge1e3/dom-id";
import { ref } from "@hoge1e3/ref";
function refList_from(raw) {
    return raw.map(ref);
}
function refList_values(raw) {
    return raw.map(e => e.value);
}
function fromDraft(d) {
    // 子要素を再帰変換して Ref<Item> にする
    const children = refList_from(d.sub.map(fromDraft));
    return {
        id: d.id,
        title: d.title,
        ok: d.ok,
        // 配列自体も ref に包む
        sub: ref(children),
    };
}
export function decode(text) {
    const lines = text.split(/\r?\n/).filter(l => l.trim().length > 0);
    const root = [];
    const stack = [{
            indent: -1, list: root
        }];
    for (let raw of lines) {
        // count leading spaces (indent)
        const indent = raw.match(/^ */)[0].length;
        let line = raw.trim();
        // detect "ok"
        let ok = false;
        if (line.startsWith("ok ")) {
            ok = true;
            line = line.slice(3);
        }
        const node = {
            id: idg().toString(),
            title: line,
            sub: []
        };
        if (ok)
            node.ok = true;
        // find correct parent by popping stack
        while (stack.length > 1 && indent <= stack[stack.length - 1].indent) {
            stack.pop();
        }
        const parent = stack[stack.length - 1].list;
        parent.push(node);
        // prepare for children
        stack.push({
            indent,
            list: node.sub
        });
    }
    // optional: remove empty sub arrays
    /*function cleanup(list){
      for (const item of list) {
        if (item.sub && item.sub.length === 0) {
          delete item.sub;
        } else if (item.sub) {
          cleanup(item.sub);
        }
      }
    }*/
    //cleanup(root);
    return refList_from(root.map(fromDraft));
}
// @acepad/todo
export function encode(data) {
    const lines = [];
    function walk(list, depth) {
        for (const item of refList_values(list)) {
            const indent = " ".repeat(depth); // 1スペース単位
            const prefix = item.ok ? "ok " : "";
            lines.push(indent + prefix + item.title);
            if (item.sub &&
                item.sub.value.length > 0) {
                walk(item.sub.value, depth + 1);
            }
        }
    }
    walk(data, 0);
    return lines.join("\n");
}
