// @acepad/todo/js/ui.js
#!runts
import { ref } from "@hoge1e3/ref";
import { decode, findById } from "./todo.js";
import { c } from "@hoge1e3/component";
import { refInput } from "@hoge1e3/ref-input";
import { create } from "@hoge1e3/immer-path-ref";
export async function main() {
    return this.widget(wmain);
}
function tap(a) {
    console.log("tap", a);
    return a;
}
function wmain({ id: wid, t }) {
    //this.echo("key");
    const root = decode(this.resolve(this.$home).rel("todo.txt").text());
    const expanded = {};
    const expandedRef = create(ref(expanded));
    const rootRef = create(ref(root));
    const showok = ref(false);
    rootRef.on("change", ({ val }) => {
        console.log("r", val);
    });
    return t.div({ style: `
  width:100%; height:100%; overflow: scroll
  ` }, refInput(showok, t.input({
        id: wid.so,
        type: "checkbox"
    })), todoListComponent(wid.root, rootRef));
    /*kstr.addEventListener("down",
    (e)=>this.echo(e.key))*/
    function todoListComponent(id, items) {
        return c(id, items, showok, (items, so) => t.ul(...items.
            filter(i => so || !i.ok).
            map(item => todoComponent(rootRef.join((root) => findById(root, item.id)), expandedRef.join(item.id)))));
    }
    function todoComponent(item, expanded) {
        const iid = wid[item.value.id];
        return c(iid, item, expanded, (itemv, expandedv) => t.li({ onclick(e) {
                console.log("ex", e.target);
                if (e?.target?.tagName.toLowerCase() !== "input")
                    expanded.value = !expanded.value;
            } }, ...[itemv.sub.length ?
                t.span("+")
                : []], refInput(item.join("ok"), t.input({ id: iid.ok, type: "checkbox" })), refInput(item.join("title"), t.input({ id: iid.title })), ...[expandedv ?
                todoListComponent(iid.sub, item.join("sub"))
                : []]));
    }
}
