// @acepad/todo/js/ui.js
#!runts
import { ref, change } from "@hoge1e3/ref";
import { decode, findById } from "./todo.js";
import { c } from "@hoge1e3/component";
import { refInput } from "@hoge1e3/ref-input";
import { fromVal } from "@hoge1e3/immer-path-ref";
//import {PathRefProxy,ReadonlyPathRefProxy,fromVal,map/*PathRef, ReadonlyPathRef*/} from "@hoge1e3/immer-path-ref/js/proxy.js";
export async function main() {
    return this.widget({
        wmain,
        css: "todo.css"
    });
}
function tap(a) {
    console.log("tap", a);
    return a;
}
function wmain({ id: wid, t }) {
    //this.echo("key");
    const root = decode(this.resolve(this.$home).rel("todo.txt").text());
    const expands = {};
    const expandsRef = fromVal(expands);
    const rootRef = fromVal(root);
    const showok = ref(false);
    const activeRef = ref(root[0].id);
    change(rootRef, ({ val }) => {
        console.log("r", val);
    });
    return t.div({ class: "scroll" }, t.div(refInput(showok, t.input({
        id: `${wid}-showok`,
        type: "checkbox"
    })), "Show ok"), todoListComponent(`${wid}-root`, rootRef));
    /*kstr.addEventListener("down",
    (e)=>this.echo(e.key))*/
    function todoListComponent(id, items) {
        return c(id, items, showok, (items, so) => t.ul(...items.
            filter(i => so || !i.ok).
            map(item => todoComponent(rootRef.join((root) => findById(root, item.id)), expandsRef.join(item.id)))));
    }
    function hasSub(i, showok) {
        if (showok)
            return i.sub.length > 0;
        return i.sub.some(child => !child.ok || hasSub(child, showok));
    }
    function op(c, f) {
        return c ? [f()] : [];
    }
    function todoComponent(itemRef, expandedRef) {
        const iid = wid + "-" + itemRef.join("id").value;
        return c(iid, itemRef, expandedRef, activeRef, (item, expanded, active) => t.li({
            class: active === item.id ?
                "active" : "",
            onclick(e) {
                console.log("ex", e.target);
                if (e?.target?.tagName.toLowerCase() !== "input") {
                    expandedRef.value = !expanded;
                }
                activeRef.value = item.id;
                e.stopPropagation();
            }
        }, ...op(hasSub(item, showok.value), () => t.span("+")), refInput(itemRef.join("ok"), t.input({ id: iid + "-ok", type: "checkbox" })), refInput(itemRef.join("title"), t.input({ id: iid + "-title" })), ...op(expanded, () => todoListComponent(iid + "-sub", itemRef.join("sub")))));
    }
}
