// @acepad/todo/js/ui.js
#!runts
import * as kstr from "@hoge1e3/key-stream";
import { computed, ref, change } from "@hoge1e3/ref";
import { decode, findById } from "./todo.js";
import { c } from "@hoge1e3/component";
import { refInput } from "@hoge1e3/ref-input";
import { fromVal } from "@hoge1e3/immer-path-ref";
import { next } from "./nav.js";
//import {PathRefProxy,ReadonlyPathRefProxy,fromVal,map/*PathRef, ReadonlyPathRef*/} from "@hoge1e3/immer-path-ref/js/proxy.js";
export async function main() {
    return this.widget({
        wmain,
        css: "todo.css"
    });
}
function wmain({ id: wid, t, on }) {
    //this.echo("key");
    const root = decode(this.resolve(this.$home).rel("todo.txt").text());
    const expands = {};
    const expandsRef = fromVal(expands);
    const rootRef = fromVal(root);
    const showok = ref(false);
    const activeRef = ref(root[0].id);
    change(rootRef, ({ val }) => {
        console.log("r", val);
    });
    on("close", () => kstr.removeEventListener("keydown", kdh));
    kstr.addEventListener("keydown", kdh);
    function kdh(e) {
        const o = { ArrowUp: -1,
            ArrowDown: 1 }[e.key];
        if (!o)
            return;
        const active = activeRef.value;
        if (!active)
            return false;
        const n = next(document.querySelector(".todo.active"), ".todo", o);
        console.log(n, active);
        if (!n)
            return false;
        activeRef.value = n.getAttribute("data-id");
        setTimeout(() => {
            document.querySelector(".todo.active .title").
                dispatchEvent(new Event("click"));
        }, 100);
        return true;
    }
    setTimeout(() => console.log("q", document.querySelectorAll(".todo")), 100);
    return t.div({ class: "scroll" }, t.div(refInput(showok, t.input({
        id: `${wid}-showok`,
        type: "checkbox"
    })), "Show ok"), todoListComponent(`${wid}-root`, rootRef));
    function todoListComponent(id, items) {
        return c(id, items, showok, (items, so) => t.ul(...items.
            filter(i => so || !i.ok).
            map(item => todoComponent(rootRef.join((root) => findById(root, item.id)), expandsRef.join(item.id)))));
    }
    function hasSub(i, showok) {
        if (showok)
            return i.sub.length > 0;
        return i.sub.some(child => !child.ok || hasSub(child, showok));
    }
    function op(c, f) {
        return c ? [f()] : [];
    }
    function todoComponent(itemRef, expandedRef) {
        const itemid = itemRef.join("id").value;
        const iid = wid + "-" + itemid;
        const isaRef = computed(activeRef, (a) => a === itemid);
        return c(iid, itemRef, expandedRef, activeRef, (item, expanded, active) => t.li({
            class: "todo " +
                (active === item.id ?
                    "active" : ""),
            "data-id": item.id,
            onclick(e) {
                console.log("ex", e.target);
                if (e?.target?.tagName.toLowerCase() !== "input") {
                    expandedRef.value = !expanded;
                }
                activeRef.value = item.id;
                e.stopPropagation();
            }
        }, ...op(hasSub(item, showok.value), () => t.span("+")), refInput(itemRef.join("ok"), t.input({ id: iid + "-ok", type: "checkbox" })), refInput(itemRef.join("title"), t.input({ id: iid + "-title", class: "title" })), ...op(expanded, () => todoListComponent(iid + "-sub", itemRef.join("sub")))));
    }
}
function tap(a) {
    console.log("tap", a);
    return a;
}
