// @acepad/browser/acepad-browser.js
import {appendNode} from "./dom.js";
import {createModuleURL, getCore} from "petit-node";
import {createURL} from "./url.js";
import {sibling,file} from "@acepad/here";
export async function open(f,_opt={}){
    let p=new DOMParser();
    let html=f.text();
    const colon=":";
    let d=p.parseFromString(html,"text/html");
    let _g=_opt.globalThis||globalThis;
    let opt={
      globalThis:_g,
        doc:_opt.doc||_opt.document||_g.document,
        resolve(p){
            return f.sibling(p);
        },
        async toURL(p,n){
            let t=n.getAttribute("type");
            if(t=="module"){
                return await createModuleURL(opt.resolve(p));
            }
            return await createURL(opt.resolve(p));
        },
        protocol(href,n){
            
        },
        open:(f)=>{
            return open(f,_opt);
        }
    };
    if(_opt.pNode){
      const g=opt.globalThis;
      const pn=await loadPNode(g, getCore());
      const {open}=await pn.importModule(file(import.meta.url));
      await open(f);
      return ;
    }
    //const s=programmableReadyState();
    //s.value="loading";
    //let h=hack();
    for(let t of ["head","body"]){
        let hd=opt.doc.querySelector(t);
        if(!_opt.keep)hd.innerHTML="";
        let hs=d.querySelector(t);
        await appendNode(hs, hd, opt);
    }
    /*h.setReadyState("interactive");
    h.setReadyState("complete");
    
    s.value="complete";
    document.dispatchEvent(new Event("readystatechange"));
    const le=new Event("load");
    window.dispatchEvent(le);*/
}

function loadPNode(g, core){
  if(g.pNode)return g.pNode;
  //console.log("loadPNode",g,core);
  const f=new g.Function ("core", `
    async function main(){
      ${pNodeHeader()};
      globalThis.pNode=pNode;
      globalThis.FS=pNode.FS;
      return pNode;
    }
    return main();
  `);
  //console.log(f+"");
  return f(core);
}
const cdn="https://cdn.jsdelivr.net/npm/";
function pNodeHeader(){
  return `${header()}
  globalThis.pNode=pNode;
  (${hackTimeouts})();
  await pNode.boot({core});
  `;
  //const FS=pNode.getFS();
  //await FS.mountAsync("/idb/","idb");`;

  function header(){
    let purl=process.env.PNODE_URL;
    // static import does not reports error
    // using relative path for import does not work in BlobURL?
    if (!purl ||
        ( !purl.match(/^https:\/\//) &&
          !purl.includes("127.0.0.1")  )
    ) purl=cdn+"petit-node@latest/dist/index.js";
    return `
const pNode=await import(${JSON.stringify(purl)});`;
  }

  function hackTimeouts(){
      for(let k of ["setTimeout","setInterval",
      "clearTimeout","clearInterval",]){
          globalThis[k]=globalThis[k].bind(globalThis);
      }
  }
}
//

function hack(){
    let h=document.addEventListener.hacked;
    h.readyState=document.readyState;
    if(!h){
        document.addEventListener.hacked=h={};
        Object.defineProperty(document,"readyState",{
            get(){
                return h.readyState;
            }
        });
        h.readystatechange=hackEventListener(
            document,"readystatechange");
        h.load=hackEventListener(
            window,"load");
        h.domcontentloaded=hackEventListener(
            document,"domcontentloaded");
        h.setReadyState=(s)=>{
            h.readyState=s;
            if(h.readyState==="interactive"){
                const e=new Event("DOMContentLoaded");
                for(let f of h.domcontentloaded){
                    f.call(document,e);
                }
            }
            if(h.readyState==="complete"){
                const e=new Event("load");
                for(let f of h.load){
                    f.call(window,e);
                }
            }
            const e=new Event("readystatechange");
            for(let f of h.readystatechange){
                f.call(document,e);
            }
        };
    }
    h.readystatechange.length=0;
    h.readystatechange.push((e)=>{
        if(typeof 
        document.onreadystatechange==="function"){
            document.onreadystatechange(e);
        }
    });
    delete document.onreadystatechange;
    h.domcontentloaded.length=0;
    h.load.length=0;
    h.load.push((e)=>{
        if(typeof 
        window.onload==="function"){
            window.onload(e);
        }
    });
    delete window.onload;
    h.readyState="loading";
    return h;
}
function hackEventListener(o,type){
    const old=o.addEventListener;
    const listeners=[];
    o.addEventListener=function (t,...a){
        if(type.toLowerCase()!==t.toLowerCase()){
            return old.apply(this,[t,...a]);
        }        
        alert(t+a);
        listeners.push(a[0]);
    };
    return listeners;
}
