// @hoge1e3/rpc/js/client.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import MutablePromise from "mutable-promise";
import { isFail, isReadyResponse, isSuccess, readyRequest } from "./types";
const debug = console.log.bind(console);
//const debug=((...args:any[])=>false);
export class Client {
    debug(...a) {
        if (this.doDebug)
            debug(`[${this.idhead}]`, ...a);
    }
    get isReady() {
        return this.readyPromise.isFulfilled;
    }
    constructor(target, channel = "default", origin = undefined, manualProbe = false) {
        this.target = target;
        this.channel = channel;
        this.origin = origin;
        this.manualProbe = manualProbe;
        this.idhead = "C" + Math.random();
        this.idseq = 1;
        this.queue = {};
        this.readyPromise = new MutablePromise();
        this.event = new EventTarget();
        this.doDebug = false;
        const t = this;
        t.idseq = 1;
        const isWorker = target instanceof Worker;
        this.receiver = (isWorker ?
            target :
            globalThis);
        if (!origin && isCrossOrigin(this.target)) {
            /*!(this.receiver instanceof Worker) &&
            typeof (this.receiver as any).importScripts!=="function" ) {*/
            throw new Error(this.channel + ": Origin must be specfied");
        }
        this.handler = (e) => {
            if (e instanceof ErrorEvent) {
                throw e.error;
            }
            this.debug("RECV", e);
            const d = e.data;
            if (d.channel !== channel)
                return;
            if (isReadyResponse(d)) {
                if (this.isReady)
                    return;
                this.debug("readyResponse is received from server!");
                this.readyPromise.resolve(true);
                return;
            }
            if (d.type == "request" || d.type === "readyRequest") {
                // Maybe from peer client. Ignore.
                return;
            }
            const q = t.queue[d.id];
            if (!q) {
                throw new Error("Missing id " + d.id + " in " + this.idhead);
            }
            this.debug("ID delete", d.id, t.queue);
            delete t.queue[d.id];
            if (isSuccess(d)) {
                q.resolve(d.result);
            }
            else if (isFail(d)) {
                q.reject(d.error);
            }
            else {
                this.debug("Invalid", d);
                throw new Error("Invalid response code");
            }
        };
        this.receiver.addEventListener("message", this.handler);
        if (!manualProbe)
            this.waitReady();
    }
    dispose() {
        this.receiver.removeEventListener("message", this.handler);
        this.event.dispatchEvent(new Event("dispose"));
    }
    addEventListener(type, handler) {
        this.event.addEventListener(type, handler);
    }
    requestReady() {
        return __awaiter(this, void 0, void 0, function* () {
            this.target.postMessage(readyRequest(this.channel), this.origin);
        });
    }
    waitReady() {
        return __awaiter(this, arguments, void 0, function* (times = Number.POSITIVE_INFINITY, duration = 50, durationMax = 1000, durationIncrement = 1) {
            for (let i = 0; i < times; i++) {
                if (this.isReady)
                    break;
                this.debug("times", i);
                this.requestReady();
                yield new Promise((s) => setTimeout(s, duration));
                if (duration < durationMax)
                    duration += durationIncrement;
            }
            if (!this.isReady) {
                throw new Error("Timeout to connect '" + this.channel + "'");
            }
        });
    }
    run(path_1) {
        return __awaiter(this, arguments, void 0, function* (path, params = {}) {
            const t = this;
            yield t.readyPromise;
            const id = t.idhead + (t.idseq++);
            t.queue[id] = new MutablePromise();
            this.debug("ADD id", id, t.queue);
            const req = {
                type: "request",
                id, path, params,
                channel: t.channel,
            };
            t.target.postMessage(req, t.origin);
            return t.queue[id];
        });
    }
}
function isCrossOrigin(target) {
    var _a;
    const m = target;
    try {
        void ((_a = m === null || m === void 0 ? void 0 : m.location) === null || _a === void 0 ? void 0 : _a.href);
        return false;
    }
    catch (e) {
        return true;
    }
}
