// @hoge1e3/rpc/js/server.js
import { isReadyRequest, isRequest, isMessegable } from "./types.js";
//const q=3;
const debug = console.log.bind(console);
//const debug=((...args:any[])=>false);
export class Server {
    debug(...a) {
        if (this.doDebug)
            debug(`[${this.idhead}]`, ...a);
    }
    constructor(...a) {
        this.idhead = "S" + Math.random();
        this.isReady = false;
        this.paths = {};
        this.event = new EventTarget();
        this.doDebug = false;
        this.target = (isMessegable(a[0]) ?
            a.shift() :
            globalThis);
        this.channel = (typeof a[0] === "string" ? a.shift() : "default");
        this.allowOrigin = a.shift();
        const channel = this.channel, allowOrigin = this.allowOrigin;
        this.handler = (e) => {
            const sendError = (_err) => {
                let err = Object.assign({ name: _err.name, message: _err.message, stack: _err.stack }, _err || {});
                try {
                    const j = JSON.stringify(err);
                    err = JSON.parse(j);
                }
                catch (je) {
                    err = err ? err.message || e + "" : "unknown";
                    this.debug("rpc:service", je, err);
                }
                respond(Object.assign(Object.assign({}, context), { error: err, status: "error", type: "fail" }));
            };
            if (e instanceof ErrorEvent) {
                sendError(e.error);
                return;
            }
            const d = e.data;
            const id = d.id;
            const respond = (m) => {
                if (e.source) {
                    // Iframe
                    e.source.postMessage(m, {
                        targetOrigin: e.origin,
                    });
                }
                else {
                    // Worker
                    this.debug("Worker respond", m);
                    this.target.postMessage(m);
                }
            };
            const context = { id, channel };
            this.debug("SERVER RECV", e, d.channel, channel);
            if (d.channel !== channel)
                return;
            if (e.origin && !allowOrigin.includes(e.origin)) {
                console.error("Invalid origin", e.origin);
                sendError(new Error(`Invalid origin: ${e.origin}`));
                return;
            }
            if (isReadyRequest(d)) {
                this.debug("Server READY", d);
                respond({ id: "READY", channel, status: "ready", type: "readyResponse" });
                return;
            }
            if (d.type == "success" || d.type === "fail" || d.type === "readyResponse") {
                // Maybe from peer server. Ignore.
                return;
            }
            if (!isRequest(d))
                throw new Error("Invalid message");
            try {
                const f = this.paths[d.path];
                if (!f)
                    return sendError(new Error("No such method: " + d.path));
                Promise.resolve(f(d.params, context)).then((r) => {
                    respond(Object.assign(Object.assign({}, context), { result: r, status: "ok", type: "success" }));
                }, sendError);
            }
            catch (ex) {
                sendError(ex);
            }
        };
        this.target.addEventListener("message", this.handler);
        this.ready();
    }
    dispose() {
        this.target.removeEventListener("message", this.handler);
        this.event.dispatchEvent(new Event("dispose"));
    }
    addEventListener(type, handler) {
        this.event.addEventListener(type, handler);
    }
    install(path, func) {
        this.paths[path] = func;
    }
    serv(path, func) {
        this.install(path, func);
    }
    ready() {
        this.isReady = true;
    }
}
;
