// @acepad/npm/deps.js
//!run
import {importModule, loadedModules, resolveEntry} from "petit-node";
//import {Class} from "@hoge1e3/recursive-safe";
import {traverse} from "@hoge1e3/dependencies";
import FS from "@hoge1e3/fs";
import {KeySet} from "@hoge1e3/keyset";

function p(...args){
    console.log(...args);
}
async function update(pa){
    const dependencies=
    [...await getDepsPkg(pa)].
    reduce((o,f)=>Object.assign(o,toDepSpec(f)),{});
    const pao=pa.obj();
    pao.dependencies=dependencies;
    save(pa,pao);
}
function toDepSpec(pa){
    const pao=pa.obj();
    return {
        [pao.name]:`^${pao.version}`
    };
}
export function toPkg(f){
    if(f.name()==="package.json")return f;
    return belongingPkg(f);
}
export async function getDepsPkg(pa,recusive=false){
    pa=toPkg(pa);
    const pao=pa.obj();
    if(recusive){
        return await traverse(pa,async (f)=>
            await getDepsPkg(f),{key:(f)=>f.path()});
    }
    const m=pa.sibling(pao.main);
    const infiles=await traverse(m,
        async (f)=>
            (await getDepsMod(f)).
            map(m=>m.file).
            filter((f)=>
                belongingPkg(f).equals(pa)));
    infiles.add(m);
    console.log("infiles",...infiles);
    const depp=new KeySet((f)=>f.path());
    for(let f of infiles){
        for(let m of await getDepsMod(f)){
            depp.add(belongingPkg(m.file));
        }
    }
    depp.delete(pa);
    return depp;
}
export function belongingPkg(f){
    f=f.isDir()?f:f.up();
    while(f){
        const p=f.rel("package.json");
        if(p.exists())return p;
        f=f.up();
    }
    return f;
}
async function getDepsMod(f){
    const m=await (resolveEntry(f).compile());
    const deps=m.dependencies;
    p(f.path(),deps);
    return deps;
}
function save(pkj,pk){
    const t=JSON.stringify(pk,null,4);
    if(pkj.text()!==t){
        pkj.text(t);
    }
}
