// @hoge1e3/ref/src/ref.ts
// @hoge1e3/ref
import {EventHandler} from "@hoge1e3/events";
export const V=Symbol.for("ref_get");
export const E=Symbol.for("ref_change");
export const F=Symbol.for("ref_fire");
const is_hoge1e3=Symbol("is_hoge1e3");
import { Remover } from "@hoge1e3/events";
export interface Ref<T> 
extends ReadonlyRef<T> {
  set(v: T): void;
  //extends RefProxy<T>
  value: T;
  [V]:T;
}
export interface RefProxy<T>
extends ReadonlyRefProxy<T>{
  [V]:T;
}

export type ChangeEvent<T>={ prev: T; val: T }
export type ChangeEventHandler<T>=(e:ChangeEvent<T>)=>void
export interface ReadonlyRef<T>
extends ReadonlyRefProxy<T>{
  get(): T;
  readonly value: T;
  on(event: "change", handler:ChangeEventHandler<T>):Remover;
  fire(event: "change", payload: ChangeEvent<T>): void;
}

export interface ReadonlyRefProxy<T>{
  readonly [V]:T;
  [E]:(h:ChangeEventHandler<T>)=>Remover;
  [F](event: "change", payload: ChangeEvent<T>): void;
}

export type RefValue<R> = 
  R extends ReadonlyRef<infer U> ?
  U : never;

export type RefValues
  <RS extends readonly ReadonlyRef<unknown>[]> = {
  [K in keyof RS]: RefValue<RS[K]>;
};

export function ref<T>(val: T): Ref<T>{
  const r=custom(
    ()=>val,
    (n)=>val=n);
  return r;
}
export function get<T>(r:RefProxy<T>):T{
  return r[V];
}
export function set<T>(r:RefProxy<T>,v:T){
  r[V]=v;
}
export function change<T>(
  r:ReadonlyRefProxy<T>,
  h:ChangeEventHandler<T>){
  r[E](h);
}

export function computed<
  RS extends readonly ReadonlyRef<unknown>[],
  T
>(
  ...args:[...RS,
  (...values:RefValues<RS>)=>T
]): ReadonlyRef<T>{
  //if(typeof a[a.length-2]!="function"){
    //args.push(()=>throw new Error())  
  //}
  let f=args.pop() as any;//(...values:RefValues<RS>)=>T;
  //if(args[])
  let val:T;
  let ti:any;
  function h(){
    const prev=val;
    val=f(...args.map((r:any)=>get(r))) as any;
    if(prev!==val){
      res.fire("change",{prev,val});
    }  
    ti=null;
  }
  const req=()=>{
    if(ti!=null)return ;
    ti=setTimeout(h,0);
  };
  for(let r of (args as any)){
    r.on("change",req);
  }
  const res=EventHandler.attachTo({
    get(){
      return val;
    },
    get value():T{
      return res.get();
    },
    get [V]():T{
      return res.get();
    },
    [E](h:ChangeEventHandler<T>){
      return res.on("change",h);
    },
    [F](event: "change", payload: ChangeEvent<T>): void{
      res.fire(event,payload)
    }
  });// as ReadonlyRef<T>;
  h();
  return res;

}
export function custom<T>(
  g: ()=>T,s:(v:T)=>void): Ref<T>{

  const res=EventHandler.attachTo({
    [is_hoge1e3]:true,
    get:g,
    set(n:T){
      const val=g();
      if(val===n)return ;
      const prev=val;
      s(n);
      res.fire("change",{
        prev,val:n,
      });
    },
    get value(){
      return g();
    },
    set value(v:T){
      this.set(v);
    },
    get [V](){
      return g();
    },
    set [V](v:T){
      this.set(v);
    },
    [E](h:ChangeEventHandler<T>){
      return res.on("change",h);
    },
    [F](event: "change", payload: ChangeEvent<T>): void{
      res.fire(event,payload)
    }
  });
  return res;
}
