// @acepad/mode/mode.js
// @acepad/mode
/*global ace*/
export function createMode(
modeName, 
{highlights:rules, getCompletions}) {
  const aceDefine = ace.define;
  rules=rules||[];
  // ユニークな HighlightRules 名
  const rulesName = `ace/mode/${modeName}_highlight_rules`;

  // ① HighlightRules
  aceDefine(
    rulesName,
    ["require", "exports", "ace/lib/oop", "ace/mode/text_highlight_rules"],
    function (require, exports) {
      const oop = require("ace/lib/oop");
      const TextHighlightRules =
        require("ace/mode/text_highlight_rules").TextHighlightRules;

      const HighlightRules = function () {
        this.$rules = {
          start: rules.map((r, i) => ({
            token: `r${modeName}${i}`,   // 内部用トークン名
            regex: r.regex
          }))
        };
      };

      oop.inherits(HighlightRules, TextHighlightRules);
      exports.HighlightRules = HighlightRules;
    }
  );

  // ② Mode
  aceDefine(
    `ace/mode/${modeName}`,
    ["require", "exports", "ace/lib/oop", "ace/mode/text", rulesName],
    function (require, exports) {
      const oop = require("ace/lib/oop");
      const TextMode = require("ace/mode/text").Mode;
      const HighlightRules = require(rulesName).HighlightRules;

      const Mode = function () {
        this.HighlightRules = HighlightRules;
        if (getCompletions) {
          this.completer = {
            getCompletions(editor, session, pos, prefix, callback) {
              try {
                Promise.resolve(
                  getCompletions(editor, session, pos, prefix)
                ).then(
                  result => callback(null, result || []),
                  err => {
                    console.error("completion error:", err);
                    callback(null, []);
                  }
                );
              } catch (e) {
                console.error("completion exception:", e);
                callback(null, []);
              }
            }
          };
        }
      };

      oop.inherits(Mode, TextMode);
      exports.Mode = Mode;
    }
  );

  // ③ CSS を動的注入
  const style = document.createElement("style");
  style.type = "text/css";

  style.textContent = rules
    .map((r, i) => 
    `.ace_r${modeName}${i} { ${r.css} }`)
    .join("\n");

  document.head.appendChild(style);
}
