// @hoge1e3/pnode-build/test/built.js
#!run
//include alias.js
export const reservedWords = {
    "break": 1, "case": 1, "catch": 1, "class": 1, "const": 1, "continue": 1,
    "debugger": 1, "default": 1, "delete": 1, "do": 1, "else": 1, "export": 1,
    "extends": 1, "false": 1, "finally": 1, "for": 1, "function": 1, "if": 1,
    "import": 1, "in": 1, "instanceof": 1, "new": 1, "null": 1, "return": 1,
    "super": 1, "switch": 1, "this": 1, "throw": 1, "true": 1, "try": 1,
    "typeof": 1, "var": 1, "void": 1, "while": 1, "with": 1, "yield": 1,
    "enum": 1, "await": 1, "implements": 1, "interface": 1, "package": 1,
    "private": 1, "protected": 1, "public": 1, "static": 1, "let": 1
};
let gbl_info;
export function jsToBlobURL(jsCodeString) {
    const blob = new Blob([jsCodeString], { type: 'application/javascript' });
    return URL.createObjectURL(blob);
}
export function uniqueName(avoidThem) {
    let theName;
    do {
        theName = "_" + Math.random().toString(36).slice(2);
    } while (avoidThem.includes(theName));
    return theName;
}
function isReservedWord(s) {
    return reservedWords[s];
}
export function valueToESCode(valueName, value, properties) {
  const keys = properties || Object.keys(value);
  const bindn = "__bind1234__";
  const bind = `const ${bindn}=(f)=>typeof f==="function" && !(f+"").startsWith("class") ? f.bind(${valueName}): f;`;
  const jsCodeString = keys.map((key) => key == "default" ?
    `export default ${valueName}.default;` :
    `export let ${isReservedWord(key) ? `_${key}` : key}=${bindn}(${valueName}.${key});`).join("\n");
  return bind + jsCodeString;
}

export function addURL(module, properties) {
  const ginf = getGlobalInfo();
  const value = module.value;
  if (value == null)
      throw new Error("Value is not set: " + module.path);
  if (module.url != null)
      throw new Error("URL is already set: " + module.path);
  const keys = properties || Object.keys(value);
  const ginfName = uniqueName(keys);
  const valueName = uniqueName([...keys, ginfName]);
  const jsCodeString = `
import ${ginfName} from "${ginf.url}";
let ${valueName}=${ginfName}.ld[${module.id}].value;
${valueToESCode(valueName, value, keys)}
`;
  let blobUrl = jsToBlobURL(jsCodeString);
  module.url = blobUrl;
  return blobUrl;
}
export function getGlobalInfo() {
    return gbl_info;
}
export async function initModuleGlobal() {
  const jsCodeString = `export default {};`;
  const blobUrl = jsToBlobURL(jsCodeString);
  const gbl = (await import(/* webpackIgnore: true */ blobUrl)).default;
  gbl_info = { 
    value: gbl, 
    url: blobUrl };
  return gbl_info;
}
await initModuleGlobal();
let ld=[];
gbl_info.value.ld=ld;
function es(path,timestamp,a){
  const dependencies=[];
  const dep=(s)=>{
    dependencies.push(ld[s]);
    if(!ld[s].url){
      addURL(ld[s]);
    }
    return ld[s].url;
  };
  const g=a.map((s)=>
    typeof s==="string"?s:dep(s)
  ).join("");
  ld.push({
    id:ld.length,
    url:jsToBlobURL(g),
  });
}
function dirname(p){
  const a=p.split("/");
  a.pop();
  return a.join("/");
}
function cjs(path,
timestamp,
dependencyMap,f){
  const exports={};
  const module={
    exports
  };
  const require=(p)=>
    ld[dependencyMap[p]].value;
  const file=path;
  const dir=dirname(file);
  f(require, exports, module, file, dir);
  ld.push({
    id:ld.length,
    value:module.exports
  });
}
/*0*/cjs("/idb/run/node_modules/@hoge1e3/pnode-build/test/d.cjs",1764333270272,{},function(require, exports, module, __filename, __dirname ){
exports.d=5;
});
/*1*/cjs("/idb/run/node_modules/@hoge1e3/pnode-build/test/c.cjs",1764333383471,{"./d.cjs":0},function(require, exports, module, __filename, __dirname ){
let d=require("./d.cjs");
exports.c=10+d.d;

});
/*2*/es("/idb/run/node_modules/@hoge1e3/pnode-build/test/a.js",1764248288067,["import {c} from /*\"./c.cjs\"*/\"",1,"\";\nexport let a=3*c;\n\n//# sourceURL=file:///idb/run/node_modules/@hoge1e3/pnode-build/test/a.js\n"]);
/*3*/es("/idb/run/node_modules/@hoge1e3/pnode-build/test/test.js",1764395056146,["import {a} from /*\"./a.js\"*/\"",2,"\";\nalert(a);\n//# sourceURL=file:///idb/run/node_modules/@hoge1e3/pnode-build/test/test.js\n"]);

import(ld.pop().url);
