// acepad-files/acepad-files.js
// acepad-files
import timer from "pause-me";
import {ok} from "assert";
import {match} from "textmatcher";
import {setMode} from "acepad-mode-shell";
import {convertStack,FS} from "acepad-os";
import {hasTrace} from "acepad-debug";
function stripPrompt(line){
    return line.replace(/^sh:\s*/,"");
}

/*global ace*/
export function openFile(sh,fn,opt){
    const acepad=sh.$acepad;
    const {getMainEditor,
    sessionInfo,createSession,
    events,locate,print,currentLine}=acepad;
    const editor=getMainEditor();
    let f;
    const move=()=>{
        //console.log("loc",opt);
        if(opt)editor.gotoLine(opt.row-0,opt.column-0);
    };
    if(typeof fn=="string"){
        f=sh.resolve(fn);
    }else{
        f=fn;
        fn=f.name();
    }
    let s=findSessionByFile(acepad,f);
    if(s){
        editor.setSession(s);
        move();
        return s;
    }
    if(f.isDir()){
        s=createDirList(sh,f);
        editor.setSession(s);
        //move();
        return s;
    }
    if(!f.exists())f.text("");
    let text=f.text();
    let t;
    s=createSession({
        text,
        file:f,
        type:"file",
        ts:f.lastUpdate(),
        val:text,
        onActivated(){
        },
        onDeactivated(){
            t.pause();
        }
    });

    let si=acepad.sessionInfo(s);
    si.on("activate",({oldSession})=>{
        if(!f.exists()){
            acepad.removeSession(s);
            let o=acepad.findSession(oldSession);
            if(o)setTimeout(()=>
            acepad.setSession(o),100);
            return ;
        }   
        t.resume();
    });
    t=timer(autoSync,100,true);
    //s=ace.createEditSession(f.text());
    editor.setSession(s);
    if(f.ext()==".html"){
        s.setMode("ace/mode/html");
    }else if(f.ext()==".php"){
        s.setMode("ace/mode/php");
    }else if(f.ext()==".c"){
        s.setMode("ace/mode/c");
    }else if(f.ext()==".py"){
        s.setMode("ace/mode/python");
    }else if(f.ext()==".json"){
        s.setMode("ace/mode/json");
    }else if(f.ext()==".css"){
        s.setMode("ace/mode/css");
    }else if(f.ext()==".tonyu"){
        s.setMode("ace/mode/tonyu");
    }else if(f.ext()==".js"||f.text().match(/^\/\/!run/)){
        s.setMode("ace/mode/javascript");
    }
    move();
    
    function autoSync(){
        let f=si.file;
        if(!f)return;
        if(f.isDir())return ;
        if(f.lastUpdate()>si.ts){
            s.setValue(
                f.text());
            si.ts=f.lastUpdate();
            si.val=s.getValue();
        }else if(si.val!=
            s.getValue()){
            f.text(s.getValue());
            si.val=s.getValue();
            si.ts=f.lastUpdate();
        }
    }
    return s;
}
export function createDirList(sh,dir){
    const acepad=sh.$acepad;
    const {getEditor,
    sessionInfo,createSession,
    events,locate,print,currentLine}=acepad;
    const resolve=(n)=>sh.resolve(n);//dir.rel(n);
    
    sh=sh.clone();
    sh.cd(dir);
    sh.outUI={
        log(...a){
            print(s,...a);    
        }
    };
    let editor;
    function onActivated(_editor){
        editor=_editor;
        ok(editor.session,"why session undefined");
        editor.session.setValue(
        dir.listFiles().
        sort((a,b)=>
            b.lastUpdate()-
            a.lastUpdate()
        ).map((p)=>p.name())
        .join("\n")+            
        "\n../"+
        "\nnew: "+
        "\nctrl-t: rename ctrl-e: remove\n"+
        dir+
        "\nsh: \n");
    }
    let s=createSession({
        type:"files",
        file:dir,
        name:dir.name(),
        onActivated,
        commands:{
            "ctrl-s":function (){
                locate(0,null);
                print("sh: ");
            },
            "ctrl-e":function (){
                locate(0,null);
                print("rm: ");
            },
            "ctrl-t":function (){
                locate(0,null);
                print("mv: ");
                locate(100,null);
                print(" to: ");
            },
            return: onEnter
        },
    });
    setMode(s,{
        sh,
        getContext(editor,session,pos,prefix){
            const line=session.getLine(pos.row);
            if(!line.match(/^sh:/))return "file";
            let picked=stripPrompt(line);
            let c=pos.column-(line.length-picked.length);
            let secondArg=picked.substring(0, c-prefix.length).match(/\s$/);
            if(secondArg)return "file";
            else return "command";
        }
    });

    function onEnter(e){
        let rm,mv;
        let f;
        //console.log("fon1",currentLine());
        let pos;
        match(currentLine(),
            /^new: *(\S+)/,
            (_,_0,n)=>{
                f=resolve(n);
                if(!f.exists()) {
                    if (n.match(/\/$/)) {
                        f.mkdir();
                    } else {
                        f.text("");
                    }
                }
                return n;
            },
            /^rm:? *(.+)/,
            (_,_0,n)=>{
                f=resolve(n);
                rm=true;
                return n;
            },
            /^mv:? *(\S+) +to: +(\S+)/,
            (_,_0,n,t)=>{
                f=resolve(n);
                mv=resolve(t);
                return n;
            },
            /^sh: *(.+)/,
            (_,_o,c)=>{
                locate.lineEnd(s);
                print(s,"\n");
                sh.exec(c).then(()=>1,(e)=>{
                    e=convertStack(e);
                    print(e.stack||e);
                });   
            }, 
            (line)=>{
                const tr=hasTrace(line);
                if(tr){
                    f=tr.file;
                    pos=tr;
                }else{
                    f=resolve(line);
                }
            });
        //console.log("fon2",f);
        
        if(!f)return ;
        if(!f.exists()){
            sh.echo("no "+f);
            return ;
        }
        if(rm){
            if(confirm("remove "+f.name()+"?")){
                f.rm();
                onActivated(editor);
            }
        }else if(mv){
            if(mv.exists()){
                alert(`${mv.name()} exists`);
            }else{
                f.moveTo(mv);
                onActivated(editor);
            }
        }else{
            openFile(sh,f,pos);
        }
    }
    return s;
    //onActivated(editor);
    //handleOnEnter(si);
}
export function findSessionByFile(acepad,f){
    return acepad.findSession((s,si)=>si.file&&
        si.file.equals(f));
}
