// @acepad/store/acepad-store.js
// acepad-store
import * as $ from "@hoge1e3/ajax";
import FS from "@hoge1e3/fs";

const hashurl=FS.getEnv("STORE_URL")||"acepad/hash.php";
const apikey=FS.getEnv("STORE_KEY");
const opt=apikey?{apikey}:{};
function $get(param){
    return $.get(hashurl,{...param,...opt});
}
function $post(param){
    return $.post(hashurl,{...param,...opt});
}
export async function put(data) { 
    let r=await $post({
        data:JSON.stringify(data)
    });
    return r.id;
}
export async function get(id) {
    return await $get({id});
}
export async function init(branch,data={}) {
    await $post({
        key:branch,
        data:JSON.stringify(data),
    });
    return await checkout(branch);
}
export async function checkout(branch,data) {
    data=data||await $get({key:branch});
    if(typeof data==="string")data={__id__:data};
    let head=data.__id__;
    return {
        data,
        async commit(data) {
            data.__prev__=head;
            let r=await $post({
                key:branch,
                data:JSON.stringify(data),
            });
            data.__id__=r.id;
            return checkout(branch, data);
        }
    };
}
