// @acepad/mode-shell/acepad-mode-shell.js
// acepad-mode-shell
export {init as initMode} from "./acepad-mode-shell2.js";
import {sessionInfo} from "@acepad/sessions";
export function completer(/*{sh,getContext}*/){
  function getContext(editor,session,pos,prefix){
    const line=session.getLine(pos.row);
    const si=sessionInfo(session);
    //console.log(si.sh===sh);
    const sh=si.sh;
    if(!line.match(/^sh:/))return "file";
    let picked=stripPrompt(line);
    let c=pos.column-(line.length-picked.length);
    let secondArg=picked.substring(0, c-prefix.length).match(/\s$/);
    if(secondArg)return "file";
    else return "command";
  }
  function stripPrompt(line){
    return line.replace(/^sh:\s*/,"");
  }

  return {
    getCompletions(editor, session, pos, prefix, callback) {
      const si=sessionInfo(session);
      const sh=si.sh;
      /*const line=editor.session.getLines(pos.row,pos.row+1)[0];
      let picked=stripPrompt(line);
      let c=pos.column-(line.length-picked.length);
      let secondArg=picked.substring(0, c-prefix.length).match(/\s$/);*/
      const sc=(w)=>w.startsWith(prefix)?1000000:10000;
      const ctx=getContext(editor,session,pos,prefix);
      switch(ctx){
      case "file":
        callback(null,(sh.getcwd().ls()).map(n=>({
          caption: n,
          value: n+" ",
          score: sc(n),
          meta: "file"
        })).filter(({score})=>score));
        break;
      case "command":
        let cl=sh.commandList();
        callback(null,cl.map(n=>({
          caption: n,
          value: n+" ",
          score: sc(n),
          meta: "command"
        })).filter(({score})=>score));
        break;
      }/*
      console.log(
        secondArg,
        picked.substring(0, c-prefix.length)+"|"+
        picked.substring(c-prefix.length, c)+"|"+
        picked.substring(c));
      callback(null,[{
        caption: "word",
        value: "wordinger",
        score: 100000,
        meta: "local"
      }]);  */
    }
  };
}
let Class;
export function setMode(session,/*{sh,getContext}*/){
  init();
  session.setMode(new Class(/*{sh,getContext}*/));
}
export function init(){
  if(Class) return Class;
  let TextMode=window.ace.require("ace/mode/text").Mode;
  Class=function (){
    this.completer=completer();
  };
  Class.prototype=new TextMode();
  return Class;
}
