// @acepad/keypad/src/keypad.ts
import {setLongtap} from './longtap.js';
import {modifierKeys,Datas,KP_ModifierKeys} from "./types.js";
import {addModifierStyle,modifierLocked,
getModifier,setModifier,clearUnlockedModifiers,
  modifierButtonPressed,modifierStateToKeyboardEvent,
  currentModifierState,
} from './modifier.js';//todo detect unused
export * as modifiers from "./modifier.js";
import {push,sync} from "@hoge1e3/key-stream";
import type {EventHandler} from "@hoge1e3/events";
//import {events} from './events.js';
//TODO
declare const $:Function;
//export let _doClick;
let events:EventHandler;
function procChr(e:HTMLElement){
    let syms=e.innerText;
    if (syms[0].toLowerCase()!==syms[0].toUpperCase()) {
        syms=syms[0].toLowerCase()+syms[0].toUpperCase()+syms.substring(1);
        $(e).empty().append(
            $("<span>").addClass("mask").addClass("no-shift").text(syms[0])
        ).append(
            $("<span>").addClass("mask").addClass("shift").text(syms[1])
        );
        if (syms[2]) {
            $(e).find(".mask").addClass("no-sym");
            $(e).append(
               $("<span>").addClass("mask").addClass("sym").text(syms[2])
            );
        }
    } else if (syms.length==2){
        $(e).empty().append(
            $("<span>").addClass("mask").addClass("no-sym").text(syms[0])
        ).append(
            $("<span>").addClass("mask").addClass("sym").text(syms[1])
        );
    }
}
export function initKey(b:HTMLElement){
    if(b.classList.contains("chr")){
        procChr(b);
    }
    setLongtap(b,{
        start(){
            doClickRender(b);
        },
        autorepeat:b.classList.contains("autorepeat")
    });
    let m=getDatas(b);//.getAttribute("data-modifier");
    if(m.modifier){
        modifierKeys[m.modifier]=1;
    }   
}
export function initKeypad({events:_events}:{events:EventHandler}){
  events=_events;
  //  _doClick=pressHandler;
    const k=document.getElementById('keypad');
    trimTextNode(k as HTMLElement);
    for(let b of document.querySelectorAll("#keypad button")) {
        initKey(b as HTMLElement);
    }
    addModifierStyle();
    justifyAll();

    //console.log()
}
export function hasSym(){
    const s=document.querySelector("[data-modifier='sym']");
    return isHTMLElement(s)&&!isHidden(s);
}
export function justifyAll(){
    $(".justify").each(function (this:HTMLElement){
        initJustify(this);
    });
}
export function initJustify(e:HTMLElement){
    const r=new ResizeObserver((ev)=>{
        justify(e);
    });
    r.observe(e);
    justify(e);
}

export function justify(row:HTMLElement){
    if(isHidden(row))return ;
    let r=[...row.childNodes].filter(
        (node)=>
        isHTMLElement(node)&&
        !isHidden(node)
    );
    let len=r.length;
    let p=Math.floor(1000/len)/10;
    for(let e of r){
        $(e).css({width:p+"%"});
    }
}
export function trimTextNode(container:Element) {
    const nodes = [...container.childNodes];
    for (const node of nodes) {
        if(!isElement(node)) continue;
        if (node.textContent.trim() === '') {
            container.removeChild(node);
        } else trimTextNode(node);
    }
}
export function isHidden(el:HTMLElement) {
    return (el.offsetParent === null);
}
function isHTMLElement(e:Node|null): e is HTMLElement{
    return !!e && e.nodeType===Node.ELEMENT_NODE;
}
function isElement(e:Node|null): e is Element {
    return !!e && (isHTMLElement(e) || e.nodeType===Node.TEXT_NODE);
}
export function getDatas(e:HTMLElement,res={} as Datas):Datas{
    if(!e.attributes)return {};
    for(let a of Array.from(e.attributes)){
        a.name.replace(/^data-(.*)/,
        (_:string,n:string)=>{
            res[n as keyof Datas]=a.value as any;
            return "";
        });
    }
    for(let c of e.childNodes){
        if (!isHTMLElement(c))continue;
        if(isHidden(c))continue;
        getDatas(c,res);
    }
    if (!res.command && !res.text && 
    !res.modifier && !res.keycode) res.text=e.innerText;
    return res;
}
export function doClickRender(b:HTMLElement) {
    events.fire("keyclick",{b});
    doClick(b);
    renderModifierState();
}
export function toggleSym(){
    if(!hasSym())return ;
    setModifier("sym",2);
    renderModifierState();
    return ;
}
export function showGuide(_b:string|HTMLElement,size=10) {
    let b:HTMLElement;
    if(typeof _b==="string"){
        b=textToButton()[_b];
    } else b=_b;
    if (!b) return;
    if(isHidden(b))return ;
    const r=b.getBoundingClientRect();
    const res=$("<span>").addClass("guide").appendTo("body");
    res.css({
        left: r.left-size,
        top: r.top-size,
        width: r.width+size*2,
        height: r.height+size,
        zIndex:300001+size,
    });
    res.text(b.innerText);
    res[0].addEventListener("touchstart",(e:TouchEvent)=>{
        doClickRender(res[0]);
        res.remove();
        e.preventDefault();
        return false;
    });
    setTimeout(shrink,200);
    function shrink() {
        if (isHidden(res)) return;
        res.css({
            left: r.left-size,
            top: r.top-size,
            width: r.width+size*2,
            height: r.height+size,
        });
        size--;
        if (size>0) setTimeout(shrink,200);
        else res.remove();
    }
    return res;
}
export function textToButton(){
    const res={} as Record<string, HTMLElement>;
    for(let b of document.querySelectorAll("#keypad button")){
        let d=getDatas(b as HTMLElement);
        let t=d.text;
        if(t)res[t]=b as HTMLElement;
    }
    return res;
}
export function renderModifierState() {
    for(let m in modifierKeys){
        if(getModifier(m as KP_ModifierKeys)){
            $("#keypad").addClass(m);
        }else{
            $("#keypad").removeClass(m);
        }
        if(modifierLocked(m as KP_ModifierKeys)){
            $("#keypad").addClass(`lock-${m}`);
        }else{
            $("#keypad").removeClass(`lock-${m}`);
        }
    }
    events.fire("renderModifierState",{});
    
}
function click(e:{key:string}){
  //console.log("click",e);
  push({type:"down",...e});
  push({type:"up",...e});
  
}
export function isArrowKeyCode(keyCode:number) {
    return keyCode>=37&&keyCode<=40;
}

export function doClick(b:HTMLElement) {
  //ok:key-stream
  // called also from acepad.doClick
  // e.g. acepad-function-keys
    //const editor=getCurrentEditor();
    let d=getDatas(b);
    let cmd=d.command;
    //let keyCode=d.keycode-0;
    let key=d.key;
    let text=d.text;
    let mod=d.modifier;
    if (getModifier("ctrl") && d.ctrl) {
      click({key:"ace:"+d.ctrl});
      clearUnlockedModifiers();
      sync(modifierStateToKeyboardEvent());
      return;
    }
    if (cmd) {
      click({key:"ace:"+cmd});
      clearUnlockedModifiers();
      sync(modifierStateToKeyboardEvent());
      return;
    }
    if (mod) {
      // should not move to keypad, 
      //if there were externally customed modifier keys,
      // doClick may be called directly.
      modifierButtonPressed(d);
      sync(modifierStateToKeyboardEvent());
      return ;
    }
    if(key&&key.match(/^Arrow/)){
        if(getModifier("select")){
          let m=Object.assign({},currentModifierState());
          m.shift=2;
          sync(modifierStateToKeyboardEvent(m));
          click({key});
          m.shift=0;
          sync(modifierStateToKeyboardEvent(m));
          return;
        }
        if(getModifier("shift")){
          setModifier("edit",2);
          setModifier("select",2);
          renderModifierState();
          
        }
    }
    if(key){
      click({key});
    }else if(text){
      click({key:text});
    }
    clearUnlockedModifiers();
    sync(modifierStateToKeyboardEvent());
}
