// @acepad/keypad/src/longtap.ts
import {setHandler} from "user-gesture";
import * as resume from "@hoge1e3/on-resume";

export function showTip(_e:TouchEvent|MouseEvent) {
    let tip=document.querySelector("#tip")! as HTMLElement;
    const b=_e.currentTarget! as HTMLElement;
    let e:{pageX:number,pageY:number}=
      (_e instanceof TouchEvent?
       _e.changedTouches[0]:_e);
    const r=b.getBoundingClientRect();
    //const rt=tip.getBoundingClientRect();
    let cx=r.left+r.width/2;
    let cy=r.top+r.height/2;
    let dx=(e.pageX-cx)/r.width;
    let dy=(e.pageY-cy)/r.height;
    let d=Math.max(Math.abs(dx),Math.abs(dy));
    //console.log(d);
    //console.log(r);
    tip.setAttribute("style",
    `display:inline;
    left:${r.left}px;
    height:${r.height}px;
    width:${r.width}px;
    top:${r.top-r.height}px;`);
    if(d>0.25){
        tip.classList.add("border");
        if (d>0.35) {
            vibrate(200);
        } else {
            vibrate(100);
        }
    }else{
        tip.classList.remove("border");
        vibrate(20);
    }
    tip.innerText=b.innerText;
}
function vibrate(t:number){
    try{
        window.navigator.vibrate(t);
    }catch(e){}
}
declare const $:Function;
export function hideTip(){
    $("#tip").hide();
}
let artimer:any;

type TapEvent=TouchEvent|MouseEvent;
type Handlers={
    longtap?: (e:TapEvent)=>void,    
    autorepeat?: (e:{currentTarget:HTMLElement})=>void,    
    start?: (e:TapEvent)=>void,    
    end?:  (e:TapEvent)=>void,
};
export function setLongtap(b:HTMLElement,handlers:Handlers){
    let dolong=handlers.longtap||
    handlers.autorepeat;
    const clearAR=()=>{
        clearTimeout(artimer);
        clearInterval(artimer);
    };
    if(handlers.autorepeat&&
    typeof handlers.autorepeat!="function"){
        handlers.autorepeat=(e)=>handlers.start(e);
    }
    setHandler(b);
    let started=false;
    const start=(e:TapEvent)=>{
        started=true;
        showTip(e);
        if (handlers.start) handlers.start(e);
        e.stopPropagation();
        e.preventDefault();
        clearAR();
        if(dolong){
            artimer=setTimeout(long,300);
        }
    };
    const long=(e:TapEvent)=>{
        if (handlers.longtap) handlers.longtap(e);
        clearAR();
        if(handlers.autorepeat){
            const ara=()=>{
                handlers.autorepeat({
                    currentTarget:b
                });
                artimer=setTimeout(ara,30);
            };
            artimer=setTimeout(ara,30);
        }
    };
    const end=(e)=>{
        hideTip(e);
        clearAR();
        if(!started)return;
        started=false;
        handlers.end({
            currentTarget:b
        });
    };
    b.addEventListener("mousedown", start);
    b.addEventListener("touchstart", start);
    b.addEventListener("mouseup", end);
    b.addEventListener("mouseleave", end);
    b.addEventListener("touchend", end);
    resume.on("resume",end);

}
