// @hoge1e3/pnode-build/tmpl-pnode.js
#!run
//include alias.js
export const reservedWords = {
    "break": 1, "case": 1, "catch": 1, "class": 1, "const": 1, "continue": 1,
    "debugger": 1, "default": 1, "delete": 1, "do": 1, "else": 1, "export": 1,
    "extends": 1, "false": 1, "finally": 1, "for": 1, "function": 1, "if": 1,
    "import": 1, "in": 1, "instanceof": 1, "new": 1, "null": 1, "return": 1,
    "super": 1, "switch": 1, "this": 1, "throw": 1, "true": 1, "try": 1,
    "typeof": 1, "var": 1, "void": 1, "while": 1, "with": 1, "yield": 1,
    "enum": 1, "await": 1, "implements": 1, "interface": 1, "package": 1,
    "private": 1, "protected": 1, "public": 1, "static": 1, "let": 1
};
let gbl_info;
export function jsToBlobURL(jsCodeString) {
    const blob = new Blob([jsCodeString], { type: 'application/javascript' });
    return URL.createObjectURL(blob);
}
export function uniqueName(avoidThem) {
    let theName;
    do {
        theName = "_" + Math.random().toString(36).slice(2);
    } while (avoidThem.includes(theName));
    return theName;
}
function isReservedWord(s) {
    return reservedWords[s];
}
export function valueToESCode(valueName, value, properties) {
    const keys = properties || Object.keys(value);
    const bindn = "__bind1234__";
    const bind = `const ${bindn}=(f)=>typeof f==="function" && !(f+"").startsWith("class") ? f.bind(${valueName}): f;`;
    const jsCodeString = keys.map((key) => key == "default" ?
        `export default ${valueName}.default;` :
        `export let ${isReservedWord(key) ? `_${key}` : key}=${bindn}(${valueName}.${key});`).join("\n");
    return bind + jsCodeString;
}

export function addURL(module, properties) {
    const ginf = getGlobalInfo();
    const value = module.value;
    if (value == null)
        throw new Error("Value is not set: " + module.path);
    if (module.url != null)
        throw new Error("URL is already set: " + module.path);
    const keys = properties || Object.keys(value);
    const ginfName = uniqueName(keys);
    const valueName = uniqueName([...keys, ginfName]);
    const jsCodeString = `
import ${ginfName} from "${ginf.url}";
let ${valueName}=${ginfName}.aliases.getByPath("${module.path}").value;
${valueToESCode(valueName, value, keys)}
`;
    let blobUrl = jsToBlobURL(jsCodeString);
    module.url = blobUrl;
    //getAliases().reload(module);
    return blobUrl;
}
export function addAlias(path, value, properties) {
  const ginf = getGlobalInfo();
  const keys = properties || Object.keys(value);
  const ginfName = uniqueName(keys);
  const valueName = uniqueName([...keys, ginfName]);
  const jsCodeString = `
import ${ginfName} from "${ginf.url}";
let ${valueName}=${ginfName}.aliases.getByPath("${path}").value;
${valueToESCode(valueName, value, keys)}
`;
  let blobUrl = jsToBlobURL(jsCodeString);
  ginf.value.aliases.set(
    path,blobUrl);
  //  new BuiltinModule(path, value, blobUrl));
}
export function getGlobalInfo() {
    return gbl_info;
}
export async function initModuleGlobal() {
    const jsCodeString = `export default {};`;
    const blobUrl = jsToBlobURL(jsCodeString);
    const gbl = (await import(/* webpackIgnore: true */ blobUrl)).default;
    gbl.aliases = new Map();
    gbl_info = { value: gbl, url: blobUrl };
    return gbl_info;
}
await initModuleGlobal();
const pNode=globalThis.pNode;
let ld=[];
let aliases=pNode?.loadedModules();
function es(path,timestamp,a){
  const dependencies=[];
  const dep=(s)=>{
    dependencies.push(ld[s]);
    if(!ld[s].url){
      addURL(ld[s]);
    }
    return ld[s].url;
  };
  const g=a.map((s)=>
    typeof s==="string"?s:dep(s)
  ).join("");
  if (pNode) {
    ld.push(pNode.addPrecompiledESModule(path, timestamp, g, dependencies));
  } else {
    ld.push({url:jsToBlobURL(g)});
  }
}
function dirname(p){
  const a=p.split("/");
  a.pop();
  return a.join("/");
}
function cjs(path,
timestamp,
dependencyMap,f){
  const exports={};
  const module={
    exports
  };
  const require=(p)=>
    ld[dependencyMap[p]].value;
  const file=path;
  const dir=dirname(file);
  f(require, exports, module, file, dir);
  ld.push({
    value:module.exports
  });
}
function b(path){
  ld.push(aliases.getByPath(path));
}
//insert
import(ld.pop().url);
