// @acepad/gsync/js/src/git.js
// @acepad/git
import { promises as fs } from 'fs';
import * as path from 'path';
import * as path_typed from "./path_typed.js";
/*import zlib from 'zlib';
import crypto from 'crypto';*/
import ignore from 'ignore';
//import { promisify } from 'util';
import { asFilename, asHash, asMode, Author, isObjectType, asBranchName, asLocalRef, asFilePath, asPathInRepo } from './types.js';
import { inflate, deflate, sha1Hex } from './codec.js';
/*const inflate = promisify(zlib.inflate);
const deflate = promisify(zlib.deflate);*/
import { getSplashScreen } from "./splash.js";
import { exists, join } from './util.js';
import { Index } from "./index_file.js";
const splashScreen = await getSplashScreen();
export class Repo {
    gitDir;
    objectStore;
    ignoreSymlink = true;
    //_objectStore:ObjectStore|undefined;
    constructor(gitDir, objectStore) {
        this.gitDir = gitDir;
        this.objectStore = objectStore;
    }
    async getObjectStore() {
        return this.objectStore;
        /*if (this._objectStore) return this._objectStore;
        this._objectStore=await factory(this.gitDir);
        return this._objectStore;*/
    }
    toFilePath(pathInRepo) {
        const r = path.join(this.workingDir(), pathInRepo);
        return asFilePath(r);
    }
    toPathInRepo(filePath) {
        const r = path.relative(this.workingDir(), filePath);
        if (r.startsWith(".."))
            throw new Error(`${filePath} is out of working dir ${this.workingDir()}`);
        return asPathInRepo(r);
    }
    async readObject(hash) {
        /*const filePath = this.getObjectPath(hash);
        const compressed = await fs.promises.readFile(filePath);*/
        const objectStore = await this.getObjectStore();
        const objval = await objectStore.get(hash);
        return Repo.objectEntryToGitObject({ hash, ...objval });
    }
    static async objectEntryToGitObject(objval) {
        const { mtime, content: compressed, hash } = objval; //await objectStore.get(hash);
        const data = Buffer.from(await inflate(new Uint8Array(compressed)));
        const nullIndex = data.indexOf(0);
        const header = data.subarray(0, nullIndex).toString();
        const [type, sizeStr] = header.split(' ');
        const size = parseInt(sizeStr, 10);
        const content = data.subarray(nullIndex + 1);
        if (content.length !== size) {
            throw new Error(`Size mismatch: expected ${size}, got ${content.length}`);
        }
        if (!isObjectType(type)) {
            throw new Error(`Unknown object type: ${type}`);
        }
        return { type, content, hash };
    }
    async writeObject(type, content) {
        const header = `${type} ${content.length}\0`;
        const store = new Uint8Array(Buffer.concat([Buffer.from(header), content]));
        const hash = asHash(await sha1Hex(store)); // crypto.createHash('sha1').update(store).digest('hex') );
        const objectStore = await this.getObjectStore();
        if (await objectStore.has(hash)) {
            return hash;
        }
        const compressed = await deflate(store);
        await objectStore.put(hash, compressed, false);
        return hash;
    }
    /*async hashObject(type: ObjectType, content: Buffer): Promise<Hash> {
      const header = `${type} ${content.length}\0`;
      const store = Buffer.concat([Buffer.from(header), content]);
      return asHash( await sha1Hex(store) );// crypto.createHash('sha1').update(store).digest('hex')
    }*/
    async readBlobAsText(hash) {
        const { type, content } = await this.readObject(hash);
        if (type !== 'blob') {
            throw new Error(`Expected blob, got ${type}`);
        }
        return content.toString('utf-8');
    }
    async readTree(hash) {
        const gito /*{ type, content }*/ = await this.readObject(hash);
        return Repo.gitObjectToTree(gito);
    }
    static async gitObjectToTree({ type, content }) {
        if (type !== 'tree') {
            throw new Error(`Expected tree, got ${type}`);
        }
        const entries = [];
        let offset = 0;
        while (offset < content.length) {
            // mode（ASCIIでspace区切り）
            const spaceIdx = content.indexOf(0x20, offset);
            const mode = asMode(content.subarray(offset, spaceIdx).toString());
            // name（null文字まで）
            const nullIdx = content.indexOf(0x00, spaceIdx);
            const name = asFilename(content.subarray(spaceIdx + 1, nullIdx).toString());
            // SHA-1（20バイトのバイナリ）
            const hashBuffer = content.subarray(nullIdx + 1, nullIdx + 21);
            const hash = asHash(hashBuffer.toString('hex'));
            entries.push({ mode, name, hash });
            offset = nullIdx + 21; // 次のエントリへ
        }
        return entries;
    }
    encodeTree(entries) {
        const buffers = [];
        for (const entry of entries) {
            const modeName = `${entry.mode} ${entry.name}`;
            const modeNameBuf = Buffer.from(modeName, 'utf8');
            const nullByte = Buffer.from([0]);
            if (entry.hash.length !== 40) {
                throw new Error(`Invalid hash length: ${entry.hash}`);
            }
            const hashBuf = Buffer.from(entry.hash, 'hex'); // 20 bytes
            buffers.push(Buffer.concat([modeNameBuf, nullByte, hashBuf]));
        }
        return Buffer.concat(buffers);
    }
    async writeTree(entries) {
        const content = this.encodeTree(entries);
        return await this.writeObject('tree', content);
    }
    async getFMTStorage() {
        return (await this.getObjectStore()).getFMTStorage();
    }
    async setHashToFMT(path, hash) {
        const fms = await this.getFMTStorage();
        if (!fms)
            return;
        const stat = await fs.lstat(path);
        if (!stat.hasFineMtime) {
            return;
        }
        const mtime = stat.mtimeMs;
        await fms.set(path, { hash, mtime });
    }
    async getHashFromFMT(path) {
        const fms = await this.getFMTStorage();
        if (!fms)
            return null;
        const stat = await fs.lstat(path);
        if (!stat.hasFineMtime) {
            await fms.clear(path);
            return null;
        }
        const mtime_current = stat.mtimeMs;
        const ent = await fms.get(path);
        if (!ent)
            return null;
        const { hash, mtime } = ent;
        if (mtime !== mtime_current) {
            await fms.clear(path);
            return null;
        }
        return hash;
    }
    async buildTreeFromWorkingDir() {
        const workingDir = this.workingDir();
        let ig = new RecursiveGitIgnore();
        const dot_gsync = path.basename(this.gitDir);
        const baseig = ignore();
        baseig.add(".git");
        baseig.add(dot_gsync);
        const walk = async (dir) => {
            ig = await ig.pushed(dir);
            try {
                const entries = [];
                const files = await fs.readdir(dir, { withFileTypes: true });
                for (const file of files) {
                    const name = asFilename(file.name);
                    const fullPath = asFilePath(path.join(dir, name));
                    const relPath = path.relative(workingDir, fullPath);
                    // 除外対象（.gitignore + .git フォルダ）をスキップ
                    if (ig.ignores(fullPath) || baseig.ignores(relPath)) {
                        //console.log("Ignored ",fullPath);
                        continue;
                    }
                    //const fullPath = asPath( path.join(dir, name) );
                    //const stat = await fs.promises.stat(fullPath);
                    if (file.isFile()) { // false for symlink
                        const _content = await fs.readFile(fullPath);
                        const content = stripCR(_content);
                        const hash = await this.writeObject('blob', content);
                        //console.log("File" , fullPath, hash);
                        entries.push({ mode: '100644', name, hash });
                    }
                    else if (file.isSymbolicLink()) {
                        if (this.ignoreSymlink)
                            continue;
                        const link = await fs.readlink(fullPath);
                        const content = Buffer.from(link, 'utf-8'); // Gitはリンク先パスをそのままblobにする
                        const hash = await this.writeObject('blob', content);
                        console.log("Symlink in tree", name, content, link);
                        entries.push({ mode: '120000', name, hash }); // ← これを追加
                    }
                    else if (file.isDirectory() && !await this.isSubRepo(fullPath)) {
                        const fmt_treeHash = await this.getHashFromFMT(fullPath);
                        if (fmt_treeHash) {
                            entries.push({ mode: '40000', name, hash: fmt_treeHash });
                        }
                        else {
                            const childEntries = await walk(fullPath);
                            const treeHash = await this.writeTree(childEntries);
                            //console.log("Dir" , fullPath, treeHash);
                            entries.push({ mode: '40000', name, hash: treeHash });
                            await this.setHashToFMT(fullPath, treeHash);
                        }
                    }
                }
                return entries;
            }
            finally {
                ig = await ig.poped();
            }
        };
        return await walk(workingDir);
    }
    workingDir() {
        return asFilePath(path.resolve(this.gitDir, '..'));
    }
    inWorkingDir(_path) {
        return path.normalize(_path).startsWith(path.normalize(this.workingDir()));
    }
    async readCommit(hash) {
        const gito /*{ type, content }*/ = await this.readObject(hash);
        return Repo.gitObjectToCommitEntry(gito);
    }
    static async gitObjectToCommitEntry({ type, content }) {
        if (type !== 'commit') {
            throw new Error(`Expected commit, got ${type}`);
        }
        const text = content.toString('utf-8');
        const lines = text.split('\n');
        /*const entry: CommitEntry = {
          tree: '',
          parents: [],
          author: '',
          committer: '',
          message: ''
        };*/
        let tree = undefined;
        let parents = [];
        let author = undefined;
        let committer = undefined;
        let message = "";
        let inMessage = false;
        const messageLines = [];
        for (const line of lines) {
            if (!inMessage && line === '') {
                inMessage = true;
                continue;
            }
            if (inMessage) {
                messageLines.push(line);
            }
            else if (line.startsWith('tree ')) {
                tree = asHash(line.slice(5));
            }
            else if (line.startsWith('parent ')) {
                parents.push(asHash(line.slice(7)));
            }
            else if (line.startsWith('author ')) {
                author = Author.parse(line.slice(7));
            }
            else if (line.startsWith('committer ')) {
                committer = Author.parse(line.slice(10));
            }
        }
        message = messageLines.join('\n');
        if (!tree)
            throw new Error("Missing tree");
        if (!author)
            throw new Error("Missing author");
        if (!committer)
            throw new Error("Missing commiter");
        return {
            tree, parents, author, committer, message
        };
    }
    encodeCommit(entry) {
        const lines = [];
        lines.push(`tree ${entry.tree}`);
        for (const parent of entry.parents) {
            lines.push(`parent ${parent}`);
        }
        lines.push(`author ${entry.author}`);
        lines.push(`committer ${entry.committer}`);
        lines.push('');
        lines.push(entry.message);
        return Buffer.from(lines.join('\n'), 'utf-8');
    }
    async writeCommit(entry) {
        return await this.writeObject("commit", this.encodeCommit(entry));
    }
    async readHead(ref) {
        const refPath = join(this.gitDir, ref);
        if (!await exists(refPath))
            return null;
        const data = await fs.readFile(refPath, 'utf-8');
        const hash = asHash(data.trim());
        return hash;
    }
    async *traverseTree(entries, prefix = '') {
        for (const entry of entries) {
            const relPath = asPathInRepo(path.posix.join(prefix, entry.name));
            const { type, content } = await this.readObject(entry.hash);
            if (type === 'blob') {
                yield { path: relPath, hash: entry.hash, content };
            }
            else if (type === 'tree') {
                yield { path: relPath, hash: entry.hash };
                const childEntries = await this.readTree(entry.hash);
                yield* this.traverseTree(childEntries, relPath); // 再帰呼び出し
            }
            else {
                // 他の型（e.g. commit）はスキップまたはエラー
                throw new Error(`Unexpected object type in tree: ${type}`);
            }
        }
    }
    async updateHead(ref, hash) {
        const fullPath = path.join(this.gitDir, ref);
        const dirPath = path.dirname(fullPath);
        await fs.mkdir(dirPath, { recursive: true });
        await fs.writeFile(fullPath, hash);
    }
    async findMergeBase(commitHashA, commitHashB) {
        // visitedA と visitedB に各ブランチの履歴を記録
        const visitedA = new Set();
        const visitedB = new Set();
        // 両方の探索キュー
        const queueA = [commitHashA];
        const queueB = [commitHashB];
        // 両方向から探索（BFS）
        while (queueA.length > 0 || queueB.length > 0) {
            // ブランチAから
            if (queueA.length > 0) {
                const hash = queueA.shift();
                if (visitedB.has(hash))
                    return hash;
                if (visitedA.has(hash))
                    continue;
                visitedA.add(hash);
                const commit = await this.readCommit(hash);
                queueA.push(...commit.parents);
            }
            // ブランチBから
            if (queueB.length > 0) {
                const hash = queueB.shift();
                if (visitedA.has(hash))
                    return hash;
                if (visitedB.has(hash))
                    continue;
                visitedB.add(hash);
                const commit = await this.readCommit(hash);
                queueB.push(...commit.parents);
            }
        }
        // 共通祖先がない（ありえないが保険）
        throw new Error(`Unrelated history: ${commitHashA} and ${commitHashB}`);
    }
    async diffTreeRecursive(oldTree, newTree, prefix = '') {
        const diffs = [];
        const oldMap = new Map(oldTree.map(e => [e.name, e]));
        const newMap = new Map(newTree.map(e => [e.name, e]));
        const names = new Set([...oldMap.keys(), ...newMap.keys()]);
        const igc = await IgnoreChecker.init(this);
        for (const name of names) {
            const oldEnt = oldMap.get(name);
            const newEnt = newMap.get(name);
            const relPath = asPathInRepo(path.posix.join(prefix, name));
            const fullPath = path_typed.join(this.workingDir(), relPath);
            if (await igc.ignores(fullPath))
                continue;
            if (oldEnt && !newEnt) {
                // 削除
                if (oldEnt.mode === '40000') {
                    const baseSub = await this.readTree(oldEnt.hash);
                    const subDiffs = await this.diffTreeRecursive(baseSub, [], relPath);
                    diffs.push(...subDiffs);
                }
                else {
                    diffs.push({ path: relPath, type: 'deleted', oldHash: oldEnt.hash, });
                }
            }
            else if (!oldEnt && newEnt) {
                // (*A)追加(same as *B)
                if (newEnt.mode === '40000') {
                    const newSub = await this.readTree(newEnt.hash);
                    const subDiffs = await this.diffTreeRecursive([], newSub, relPath);
                    diffs.push(...subDiffs);
                }
                else {
                    diffs.push({ path: relPath, type: 'added', newHash: newEnt.hash, newMode: newEnt.mode });
                }
            }
            else if (oldEnt && newEnt) {
                if (oldEnt.mode === '40000' && newEnt.mode === '40000') {
                    // ディレクトリどうし → 再帰
                    const oldSub = await this.readTree(oldEnt.hash);
                    const newSub = await this.readTree(newEnt.hash);
                    const subDiffs = await this.diffTreeRecursive(oldSub, newSub, relPath);
                    diffs.push(...subDiffs);
                }
                else if (oldEnt.hash !== newEnt.hash) {
                    if (oldEnt.mode === '100644' && newEnt.mode === '100644') {
                        // 通常ファイルが変更
                        diffs.push({
                            path: relPath,
                            type: 'modified',
                            oldHash: oldEnt.hash,
                            newHash: newEnt.hash,
                        });
                    }
                    else {
                        // 種類が違う場合、削除→追加
                        if (oldEnt.mode === '40000') {
                            const oldSub = await this.readTree(oldEnt.hash);
                            const subDiffs = await this.diffTreeRecursive(oldSub, [], relPath);
                            diffs.push(...subDiffs);
                        }
                        else {
                            diffs.push({ type: "deleted", path: relPath, oldHash: oldEnt.hash });
                        }
                        // (*B)追加(same as *A)
                        if (newEnt.mode === '40000') {
                            const newSub = await this.readTree(newEnt.hash);
                            const subDiffs = await this.diffTreeRecursive([], newSub, relPath);
                            diffs.push(...subDiffs);
                        }
                        else {
                            diffs.push({ path: relPath, type: 'added', newHash: newEnt.hash, newMode: newEnt.mode });
                        }
                    }
                }
                // else: hash が同じ → 無視
            }
        }
        return diffs;
    }
    async threeWayMerge(baseTree, aTree, bTree) {
        const [diffA, diffB] = await Promise.all([
            this.diffTreeRecursive(baseTree, aTree),
            this.diffTreeRecursive(baseTree, bTree)
        ]);
        const toA = [];
        const toB = [];
        const conflicts = [];
        const allPaths = new Set([...diffA, ...diffB].map(d => d.path));
        const diffMapA = new Map(diffA.map(d => [d.path, d]));
        const diffMapB = new Map(diffB.map(d => [d.path, d]));
        for (const path of allPaths) {
            const da = diffMapA.get(path);
            const db = diffMapB.get(path);
            if (da && db) {
                if (da.type === "deleted" && db.type === "deleted") {
                }
                else if (da.type === "deleted") {
                    toA.push(db);
                }
                else if (db.type === "deleted") {
                    toB.push(da);
                }
                else if (da.type === "modified" && db.type === "modified") {
                    if (da.oldHash !== db.oldHash) {
                        throw new Error(`old hash does not match ${da.oldHash} !== ${db.oldHash}`);
                    }
                    conflicts.push({ path, base: da.oldHash, a: da.newHash, b: db.newHash });
                }
                else if (da.type === "added" && db.type === "added") {
                    conflicts.push({ path, a: da.newHash, b: db.newHash });
                }
                else {
                    // add & modified never happens
                    throw new Error(`Invalid state: a: ${da.type}, b: ${db.type}`);
                }
            }
            else if (db) {
                toA.push(db);
            }
            else if (da) {
                toB.push(da);
            }
        }
        return {
            toA, toB, conflicts
        };
    }
    async checkoutTreeToDir(treeHash, dirPath) {
        const entries = await this.readTree(treeHash);
        for (const entry of entries) {
            const outPath = asFilePath(path.join(dirPath, entry.name));
            await splashScreen.show(outPath);
            if (entry.mode === '40000') {
                // ディレクトリ（tree）→ 再帰的に処理
                await fs.mkdir(outPath, { recursive: true });
                await this.checkoutTreeToDir(entry.hash, outPath);
            }
            else if (entry.mode === '120000') {
                if (this.ignoreSymlink)
                    continue;
                const { type, content } = await this.readObject(entry.hash);
                if (type !== 'blob') {
                    throw new Error(`Unexpected object type ${type} for symlink ${entry.name}`);
                }
                const link = content.toString('utf-8');
                console.log("checked out Symlink", link, outPath);
                await fs.symlink(link, outPath, "junction");
            }
            else {
                // blob → ファイルとして書き出す
                const { type, content } = await this.readObject(entry.hash);
                if (type !== 'blob') {
                    throw new Error(`Unexpected object type ${type} for ${entry.name}`);
                }
                await fs.writeFile(outPath, content);
            }
        }
    }
    async getCurrentBranchName() {
        const headPath = this.headPath();
        const content = await fs.readFile(headPath, 'utf8');
        const match = content.match(/^ref: refs\/heads\/(.+)\s*$/);
        if (match) {
            return asBranchName(match[1]); // 例: "main"
        }
        else {
            // detached HEAD の場合（SHA-1直書き）
            throw new Error("detached HEAD!!");
            //return null;
        }
    }
    headPath() {
        return join(this.gitDir, 'HEAD');
    }
    async setCurrentBranchName(branch) {
        const headPath = this.headPath();
        const refPath = asLocalRef(branch);
        const content = `ref: ${refPath}\n`;
        await fs.writeFile(headPath, content, 'utf8');
    }
    async readMergeHead() {
        const MERGE_HEAD = join(this.gitDir, "MERGE_HEAD");
        if (!await exists(MERGE_HEAD))
            return null;
        return asHash(await fs.readFile(MERGE_HEAD, { encoding: "utf-8" }));
    }
    async writeMergeHead(commitHash) {
        const MERGE_HEAD = join(this.gitDir, "MERGE_HEAD");
        if (commitHash) {
            await fs.writeFile(MERGE_HEAD, commitHash);
        }
        else {
            if (await exists(MERGE_HEAD))
                await fs.rm(MERGE_HEAD);
        }
    }
    realGitRepoIsSubRepo() {
        return false; // TODO: configure
    }
    async isSubRepo(dir) {
        if (!this.inWorkingDir(dir))
            return false;
        const dot_gsync = path.basename(this.gitDir);
        return (this.realGitRepoIsSubRepo() && await exists(join(dir, ".git"))) ||
            await exists(join(dir, dot_gsync));
    }
    async inSubRepo(_path) {
        if (!this.inWorkingDir(_path))
            return false;
        const workDir = this.workingDir();
        for (; path.normalize(workDir) !== path.normalize(_path); _path = asFilePath(path.dirname(_path))) {
            if (await this.isSubRepo(_path))
                return true;
        }
        return false;
    }
    async applyDiff(diffs) {
        const workDir = this.workingDir();
        const igc = await IgnoreChecker.init(this);
        for (const diff of diffs) {
            const filePath = asFilePath(path.join(workDir, diff.path));
            const filePathParent = asFilePath(path.dirname(filePath));
            // do not skip symlink itself
            if (await this.hasSymlinkInPath(filePathParent)) {
                continue;
            }
            if (await igc.ignores(filePath))
                continue;
            if (await this.inSubRepo(asFilePath(path.dirname(filePath))))
                continue;
            if (diff.type === 'deleted') {
                if (await exists(filePath))
                    await fs.rm(filePath, { force: true });
            }
            else if (diff.type === 'added' || diff.type === 'modified') {
                await splashScreen.show("Write " + filePath);
                if (!diff.newHash)
                    throw new Error(`Missing 'other' hash for ${diff.path}`);
                const { type, content } = await this.readObject(diff.newHash);
                if (type !== 'blob')
                    throw new Error(`Expected blob, got ${type} for ${diff.path}`);
                if (diff.type === 'added' && diff.newMode === "120000") {
                    if (this.ignoreSymlink)
                        continue;
                    const linkTarget = content.toString('utf-8'); // ★必須
                    console.log("Writing symlink:", {
                        target: linkTarget,
                        path: filePath
                    });
                    await fs.symlink(linkTarget, filePath);
                    //console.log("Writing symlink into ",filePath);
                    //await fs.promises.symlink(content, filePath, "junction");
                }
                else {
                    await writeFileIgnoreingCRLF(filePath, content);
                }
            }
        }
    }
    async hasSymlinkInPath(targetPath) {
        const rel = path.relative(this.workingDir(), targetPath);
        const parts = rel.split(path.sep);
        let current = this.workingDir();
        for (const part of parts) {
            if (!part)
                continue;
            if (part == "..")
                throw new Error(`${rel}(${targetPath}) is out of this repo.`);
            current = asFilePath(path.join(current, part));
            try {
                const stat = await fs.lstat(current);
                if (stat.isSymbolicLink()) {
                    return true;
                }
            }
            catch (e) {
                // 存在しないパスの場合は「ない」と判断
                return false;
            }
        }
        return false;
    }
    async updateIndexFromWorkingDir(indexPath) {
        const workingDir = this.workingDir();
        const index = await Index.read(indexPath);
        const scannedPaths = new Set();
        let ig = new RecursiveGitIgnore();
        const dot_gsync = path.basename(this.gitDir);
        const baseig = ignore();
        baseig.add(".git");
        baseig.add(dot_gsync);
        const walk = async (dir) => {
            ig = await ig.pushed(dir);
            try {
                const files = await fs.readdir(dir, { withFileTypes: true });
                for (const file of files) {
                    const name = asFilename(file.name);
                    const fullPath = asFilePath(path.join(dir, name));
                    const relPath = path.relative(workingDir, fullPath);
                    // Ignore checker
                    if (ig.ignores(fullPath) || baseig.ignores(relPath)) {
                        continue;
                    }
                    const relPathSlash = relPath.replace(/\\/g, "/");
                    if (file.isFile() || file.isSymbolicLink()) {
                        const stat = await fs.lstat(fullPath);
                        scannedPaths.add(relPathSlash);
                        // Check if index already has a valid entry with matching mtime and size
                        const existing = index.entries.get(relPathSlash);
                        const mtimeMs = stat.mtime.getTime();
                        const mtimeSec = Math.floor(mtimeMs / 1000);
                        const mtimeNano = (mtimeMs % 1000) * 1000000;
                        if (existing && existing.mtimeSec === mtimeSec && existing.fileSize === stat.size) {
                            continue;
                        }
                        let hash;
                        let mode = 0o100644; // 33188
                        if (file.isSymbolicLink()) {
                            if (this.ignoreSymlink)
                                continue;
                            const link = await fs.readlink(fullPath);
                            const content = Buffer.from(link, 'utf-8');
                            hash = await this.writeObject('blob', content);
                            mode = 0o120000; // 40960
                        }
                        else {
                            const _content = await fs.readFile(fullPath);
                            const content = stripCR(_content);
                            hash = await this.writeObject('blob', content);
                        }
                        const ctimeMs = stat.ctime.getTime();
                        const ctimeSec = Math.floor(ctimeMs / 1000);
                        const ctimeNano = (ctimeMs % 1000) * 1000000;
                        index.entries.set(relPathSlash, {
                            ctimeSec,
                            ctimeNano,
                            mtimeSec,
                            mtimeNano,
                            dev: stat.dev || 0,
                            ino: stat.ino || 0,
                            mode,
                            uid: stat.uid || 0,
                            gid: stat.gid || 0,
                            fileSize: stat.size,
                            sha1: hash,
                            flags: relPathSlash.length & 0xFFF,
                            path: relPathSlash
                        });
                    }
                    else if (file.isDirectory() && !await this.isSubRepo(fullPath)) {
                        await walk(fullPath);
                    }
                }
            }
            finally {
                ig = await ig.poped();
            }
        };
        await walk(workingDir);
        // Remove deleted files
        for (const key of index.entries.keys()) {
            if (!scannedPaths.has(key)) {
                index.entries.delete(key);
            }
        }
        await index.write(indexPath);
        return index;
    }
    async writeTreeFromIndex(index) {
        const root = { children: new Map() };
        for (const entry of index.entries.values()) {
            const parts = entry.path.split("/");
            let current = root;
            for (let i = 0; i < parts.length; i++) {
                const part = parts[i];
                if (!current.children.has(part)) {
                    current.children.set(part, { children: new Map() });
                }
                current = current.children.get(part);
                if (i === parts.length - 1) {
                    current.hash = entry.sha1;
                    const octalMode = entry.mode.toString(8);
                    current.mode = asMode(octalMode);
                }
            }
        }
        const writeNodeTree = async (node) => {
            const entries = [];
            for (const [name, child] of node.children.entries()) {
                if (child.children.size > 0) {
                    // Directory
                    const dirHash = await writeNodeTree(child);
                    entries.push({
                        mode: "40000",
                        name: asFilename(name),
                        hash: dirHash
                    });
                }
                else {
                    // File or Symlink
                    if (!child.hash || !child.mode) {
                        throw new Error(`Invalid leaf node structure for ${name}`);
                    }
                    entries.push({
                        mode: child.mode,
                        name: asFilename(name),
                        hash: child.hash
                    });
                }
            }
            // Sort entries according to Git tree rules
            entries.sort((a, b) => {
                const nameA = a.mode === "40000" ? a.name + "/" : a.name;
                const nameB = b.mode === "40000" ? b.name + "/" : b.name;
                return nameA < nameB ? -1 : nameA > nameB ? 1 : 0;
            });
            return await this.writeTree(entries);
        };
        return await writeNodeTree(root);
    }
    async getBranches() {
        const headsDir = path.join(this.gitDir, 'refs', 'heads');
        const branches = [];
        const walk = async (dir) => {
            if (!await exists(dir))
                return;
            const entries = await fs.readdir(dir, { withFileTypes: true });
            for (const entry of entries) {
                const fullPath = asFilePath(path.join(dir, entry.name));
                if (entry.isDirectory()) {
                    await walk(fullPath);
                }
                else if (entry.isFile()) {
                    const rel = path.relative(headsDir, fullPath);
                    const branchName = rel.replace(/\\/g, '/');
                    branches.push(asBranchName(branchName));
                }
            }
        };
        await walk(asFilePath(headsDir));
        return branches;
    }
    async hasUncommittedChanges() {
        const branch = await this.getCurrentBranchName();
        const ref = asLocalRef(branch);
        const curCommitHash = await this.readHead(ref);
        if (!curCommitHash)
            return false;
        const curCommit = await this.readCommit(curCommitHash);
        const saveIndexFile = this.objectStore.useIndexFile();
        let newCommitTreeHash;
        if (saveIndexFile) {
            const indexPath = asFilePath(path.join(this.gitDir, "index"));
            const index = await this.updateIndexFromWorkingDir(indexPath);
            newCommitTreeHash = await this.writeTreeFromIndex(index);
        }
        else {
            const tree = await this.buildTreeFromWorkingDir();
            newCommitTreeHash = await this.writeTree(tree);
        }
        const MERGE_HEAD = await this.readMergeHead();
        return !!MERGE_HEAD || curCommit.tree !== newCommitTreeHash;
    }
}
export async function writeFileIgnoreingCRLF(filePath, content) {
    if (await exists(filePath)) {
        const org = await fs.readFile(filePath);
        if (sameExceptCRLF(org, content))
            return;
    }
    await fs.mkdir(path.dirname(filePath), { recursive: true });
    await fs.writeFile(filePath, content);
}
export class IgnoreChecker {
    repo;
    igs = new Map;
    baseig;
    static async init(repo) {
        const res = new IgnoreChecker(repo);
        const workingDir = repo.workingDir();
        res.igs.set(workingDir, await (new RecursiveGitIgnore().pushed(workingDir)));
        return res;
    }
    constructor(repo) {
        this.repo = repo;
        //const workingDir = repo.workingDir();
        //this.igs.set(workingDir,new RecursiveGitIgnore().pushed(workingDir));
        const dot_gsync = path.basename(repo.gitDir);
        this.baseig = ignore();
        this.baseig.add(".git");
        this.baseig.add(dot_gsync);
    }
    async get(dir) {
        if (!this.repo.inWorkingDir(dir))
            throw new Error(`${dir} is out of working dir ${this.repo.workingDir()}`);
        if (this.igs.has(dir)) {
            return this.igs.get(dir);
        }
        const parent = path_typed.dirname(dir);
        const r = await (await this.get(parent)).pushed(dir);
        this.igs.set(dir, r);
        return r;
    }
    async ignores(file) {
        const rel = path.relative(this.repo.workingDir(), file);
        if (this.baseig.ignores(rel))
            return true;
        const dir = path_typed.dirname(file);
        const ig = await this.get(dir);
        return ig.ignores(file);
    }
}
export class RecursiveGitIgnore {
    stack;
    constructor(stack = []) {
        this.stack = stack;
    }
    ignores(file) {
        for (const { ig, dir } of this.stack) {
            if (file.startsWith(dir)) {
                const rel = path.relative(dir, file);
                //console.log("ignores", dir, rel);
                if (rel !== "" && ig.ignores(rel))
                    return true;
            }
        }
        return false; // デフォルトは無視しない
    }
    async pushed(dir) {
        const ig = ignore();
        // .gitignore を読み込む（存在する場合）
        const gitignorePath = path.join(dir, '.gitignore');
        try {
            const ignoreContent = await fs.readFile(gitignorePath, 'utf8');
            ig.add(ignoreContent);
        }
        catch {
            // .gitignore がない場合は無視
        }
        return new RecursiveGitIgnore([...this.stack, { ig, dir }]);
        //this.stack.push({ig, dir});
    }
    async poped() {
        return new RecursiveGitIgnore(this.stack.slice(0, this.stack.length - 1));
        //this.stack.pop();
    }
}
export function stripCR(content) {
    function isUtf8Text(buffer) {
        try {
            const text = new TextDecoder().decode(buffer);
            // そもそも\r\nがないのだったら置換されないのでnull(そのまま返す)
            if (!text.includes("\r\n"))
                return null;
            // 制御文字（タブ、改行、キャリッジリターンは許可）
            const controlChars = /[\x00-\x08\x0B\x0C\x0E-\x1F\x7F]/;
            if (controlChars.test(text)) {
                return null;
            }
            // 文字化けの典型：置換文字（黒いひし形の中に?）が多く含まれていたらNG
            const replacementCharCount = (text.match(/\uFFFD/g) || []).length;
            const replacementRate = replacementCharCount / text.length;
            if (replacementRate > 0.01) {
                return null;
            }
            return text;
        }
        catch (e) {
            // .toString('utf8')で例外が起きることはまれだが念のため
            return null;
        }
    }
    const t = isUtf8Text(content);
    if (!t)
        return content;
    return new TextEncoder().encode(t.replace(/\r\n/g, "\n"));
}
export function sameExceptCRLF(a, b) {
    const sa = stripCR(a);
    const sb = stripCR(b);
    if (sa.byteLength !== sb.byteLength)
        return false;
    for (let i = 0; i < sa.byteLength; i++)
        if (sa[i] !== sb[i])
            return false;
    return true;
}
