// @acepad/gsync/js/src/util.js
import path from "path";
import { asFilePath, asPHPTimestamp } from "./types.js";
import { promises as fs } from "fs";
export function toBase64(content) {
    const ca = content;
    if (typeof ca.toBase64 === "function")
        return ca.toBase64();
    if (typeof btoa !== "undefined") {
        let binary = "";
        for (let i = 0; i < content.length; i++) {
            binary += String.fromCharCode(content[i]);
        }
        return btoa(binary);
    }
    else {
        // Node.js environment without Buffer (fallback)
        throw new Error("Base64 encoding not supported in this environment without Buffer.");
    }
}
export function phpTimestampToDate(phpts) {
    return new Date(phpts * 1000);
}
export function dateToPhpTimestamp(d) {
    return asPHPTimestamp(Math.floor(d.getTime() / 1000));
}
export function join(f, ...s) {
    return asFilePath(path.join(f, ...s));
}
export function dirname(f) {
    return asFilePath(path.dirname(f));
}
export async function exists(f) {
    try {
        await fs.access(f);
        return true;
    }
    catch (e) {
        return false;
    }
}
