// acepad-os/index.js
/*global globalThis*/
import {} from "./loadModule.js";
const o=globalThis.FS.os;
export const createModuleURL=o.createModuleURL;
export const loadModule=o.loadModule;
export const importModule=o.importModule;
//export const createURL=o.createURL;
//export const loadScriptTag=o.loadScriptTag;
export const FS=globalThis.FS;
export const convertStack=(e)=>{
    if(e && typeof e.stack==="string"){
        e.stack=o.convertStack(e.stack);
        return e;
    }
    return o.convertStack(e);
}    ;    
export const resolveModule=o.resolveModule;
function urlToPath(url){
    let e=globalThis.urls.get({url});
    return e && e.path;
}
FS.os.urlToPath=FS.os.urlToPath||urlToPath;

export function loadScriptTag(url,attr){
/*global define,requirejs*/
    if (attr.type!=="module" && 
        typeof define==="function" && 
        define.amd && 
        typeof requirejs==="function") {
        return new Promise(
            (s)=>requirejs([url],(r)=>s(r)));
    }
    const script = document.createElement('script');
    script.src = url;
    for(let k in attr){
        script.setAttribute(k,attr[k]);
    }
    return new Promise(
        function (resolve,reject){
            script.addEventListener("load",resolve);
            script.addEventListener("error",reject);
            document.head.appendChild(script);
    });
}
