// acepad-with-fs/acepad-with-fs.js
// acepad-with-fs
import {open} from "acepad-browser";
import {trial,debugSession} from "acepad-debug";
import {showWidget} from "acepad-widget";
import {initVConsole,showVConsole} from "show-vconsole";
import {run} from "acepad-find";
import {add as addLog} from "./logging.js";
import unload from "@hoge1e3/unload";
import {Menu} from "acepad-menu";
/*global $*/
export async function main(){
    const sh=this;
    await open(sh.resolve(__dirname).rel("ace.html"));
    const {setForce}=unload({
        oncancel(){
           setTimeout(()=> 
          document.documentElement.requestFullscreen().then(
              ()=>document.exitFullscreen())
        ,100);  
        }
    });
    let acepad=window.acepad=await 
    (await sh.jsm("acepad")).init();   
    sh.$acepad=window.acepad;
    let {openFile}=await sh.jsm("acepad-files");
    acepad.attachCommands({
        "ctrl-r":()=>location.reload(),
        showDebug:{bindKey:"ctrl-d",exec(){
            acepad.changeSession("*debug*");
        }},
        find:{
            bindKey:"ctrl-f",
            exec(){
                run(sh);
            }
        }
    });
    async function ngram(){
        for(let f of sh.resolve(".").recursive()){
            if(f.ext()!==".js")continue;
            let s=f.text();
            acepad.addToNgram(s);
            await 0;
        }
    }
    //ngram();
    initVConsole();
    /*acepad.showMenuButtons({
        cons(){
            showVConsole();
        },
        
    });*/
    acepad.events.on("keyclick",({b})=>{
        if(acepad.getCurrentEditor().nolog)return ;
        addLog(sh,{text:b.innerText});
        //console.log("k",b.innerText);
    });
    initMenuButtons(acepad);
    openFile(sh,"./");
    debugSession(sh);
    sh.ngr2();
    sh.filewatch();
    sh.polyfiller();
   // sh.jsm("@hoge1e3/no-obj-obj");
}
let prevs;
export function initMenuButtons(acepad){
    let {events,sessionInfo,changeSession}=acepad;
    let menu=Menu(acepad.getMainEditor());
    events.on("createSession",({name,session})=>{
        //console.log(name);
        let m;
        let si=sessionInfo(session);
        si.on("infochanged",(d)=>{
            m.text(d.name);
        });
        m=menu.add(name,()=>{
            changeSession(session);
        }); 
        si.on("remove",()=>{
            m.remove();
        });
        m.go2nd();
        menu.sort();
        si.menuButton=m;
    });
    events.on("changeSession",({oldSession,session})=>{
        let si=sessionInfo(session);
        let osi=sessionInfo(oldSession);
        if(!si.menuButton||!osi.menuButton)return ;
        $(osi.menuButton.element).removeClass("selected");
        $(si.menuButton.element).addClass("selected");
        si.menuButton.setOrder(0);
        osi.menuButton.setOrder(0);
        
        menu.sort();
    });
    menu.add({
        cons(){
            showVConsole();
        },
        ses(){
            acepad.changeSession("*sessions*");
        },
        gutt(){
            const editor=acepad.getCurrentEditor();
            editor.renderer.setShowGutter(
                !editor.renderer.getShowGutter());
        },
    });
}
