// @hoge1e3/key-stream/key-stream.js
#!run
// @hoge1e3/key-stream
/*global KeyboardEvent, EventTarget*/
const by={
  key:new Map(),
  code:new Map(),
  keyCode:new Map(),
};
const k=`{"key":"a","code":"KeyA","keyCode":65} 
{"key":"Escape","code":"Escape","keyCode":27}
{"key":"F1","code":"F1","keyCode":112}
{"key":"F2","code":"F2","keyCode":113}
{"key":"F3","code":"F3","keyCode":114}
{"key":"F4","code":"F4","keyCode":115}
{"key":"F5","code":"F5","keyCode":116}
{"key":"F6","code":"F6","keyCode":117}
{"key":"F7","code":"F7","keyCode":118}
{"key":"F8","code":"F8","keyCode":119}
{"key":"F9","code":"F9","keyCode":120}
{"key":"F10","code":"F10","keyCode":121}
{"key":"F11","code":"F11","keyCode":122}
{"key":"F12","code":"F12","keyCode":123}
{"key":"Pause","code":"Pause","keyCode":19}
{"key":"Escape","code":"Escape","keyCode":27}
{"key":"Insert","code":"Insert","keyCode":45}
{"key":"Delete","code":"Delete","keyCode":46}
{"key":"Zenkaku","code":"Backquote","keyCode":244}
{"key":"1","code":"Digit1","keyCode":49}
{"key":"2","code":"Digit2","keyCode":50}
{"key":"3","code":"Digit3","keyCode":51}
{"key":"4","code":"Digit4","keyCode":52}
{"key":"5","code":"Digit5","keyCode":53}
{"key":"6","code":"Digit6","keyCode":54}
{"key":"7","code":"Digit7","keyCode":55}
{"key":"8","code":"Digit8","keyCode":56}
{"key":"9","code":"Digit9","keyCode":57}
{"key":"0","code":"Digit0","keyCode":48}
{"key":"-","code":"Minus","keyCode":189}
{"key":"^","code":"Equal","keyCode":222}
{"key":"\\\\","code":"IntlYen","keyCode":220}
{"key":"Backspace","code":"Backspace","keyCode":8}
{"key":"Tab","code":"Tab","keyCode":9}
{"key":"q","code":"KeyQ","keyCode":81}
{"key":"w","code":"KeyW","keyCode":87}
{"key":"e","code":"KeyE","keyCode":69}
{"key":"r","code":"KeyR","keyCode":82}
{"key":"t","code":"KeyT","keyCode":84}
{"key":"y","code":"KeyY","keyCode":89}
{"key":"u","code":"KeyU","keyCode":85}
{"key":"i","code":"KeyI","keyCode":73}
{"key":"o","code":"KeyO","keyCode":79}
{"key":"p","code":"KeyP","keyCode":80}
{"key":"@","code":"BracketLeft","keyCode":192}
{"key":"[","code":"BracketRight","keyCode":219}
{"key":"Enter","code":"Enter","keyCode":13}
{"key":"Control","code":"ControlLeft","keyCode":17}
{"key":"a","code":"KeyA","keyCode":65}
{"key":"s","code":"KeyS","keyCode":83}
{"key":"d","code":"KeyD","keyCode":68}
{"key":"f","code":"KeyF","keyCode":70}
{"key":"g","code":"KeyG","keyCode":71}
{"key":"h","code":"KeyH","keyCode":72}
{"key":"j","code":"KeyJ","keyCode":74}
{"key":"k","code":"KeyK","keyCode":75}
{"key":"l","code":"KeyL","keyCode":76}
{"key":";","code":"Semicolon","keyCode":187}
{"key":":","code":"Quote","keyCode":186}
{"key":"]","code":"Backslash","keyCode":221}
{"key":"Shift","code":"ShiftLeft","keyCode":16}
{"key":"z","code":"KeyZ","keyCode":90}
{"key":"x","code":"KeyX","keyCode":88}
{"key":"c","code":"KeyC","keyCode":67}
{"key":"v","code":"KeyV","keyCode":86}
{"key":"b","code":"KeyB","keyCode":66}
{"key":"n","code":"KeyN","keyCode":78}
{"key":"m","code":"KeyM","keyCode":77}
{"key":",","code":"Comma","keyCode":188}
{"key":".","code":"Period","keyCode":190}
{"key":"/","code":"Slash","keyCode":191}
{"key":"\\\\","code":"IntlRo","keyCode":226}
{"key":"Shift","code":"","keyCode":16}
{"key":"Control","code":"ControlLeft","keyCode":17}
{"key":"Control","code":"ControlLeft","keyCode":17}
{"key":"a","code":"KeyA","keyCode":65}
{"key":"Control","code":"ControlLeft","keyCode":17}
{"key":"Meta","code":"MetaLeft","keyCode":91}
{"key":"Meta","code":"MetaLeft","keyCode":91}
{"key":"Alt","code":"AltLeft","keyCode":18}
{"key":" ","code":"Space","keyCode":32}
{"key":"ContextMenu","code":"ContextMenu","keyCode":93}
{"key":"Control","code":"ControlRight","keyCode":17}
{"key":"ArrowUp","code":"ArrowUp","keyCode":38}
{"key":"ArrowDown","code":"ArrowDown","keyCode":40}
{"key":"ArrowLeft","code":"ArrowLeft","keyCode":37}
{"key":"ArrowRight","code":"ArrowRight","keyCode":39}
{"key":"Shift","code":"ShiftLeft","keyCode":16}
{"key":"A","code":"KeyA","keyCode":65}
{"key":"C","code":"KeyC","keyCode":67}
{"key":"Shift","code":"ShiftLeft","keyCode":16}
{"key":"V","code":"KeyV","keyCode":86}
{"key":"Shift","code":"","keyCode":16}
{"key":"+","code":"Semicolon","keyCode":187}
{"key":"*","code":"Quote","keyCode":186}
{"key":"}","code":"Backslash","keyCode":221}`.
split("\n").map((l)=>{
  //console.log(l);
  return JSON.parse(l);
});
for(let n of k){
  for(let e in by){
    by[e].set(n[e],n);
  }
}
/*
Escape F1 F2 F3 F4 F5 F6 F7 F8 F9 F10 F11 F12 
Pause Insert Delete Zenkaku Backspace Tab Enter 
Control Shift Meta Alt ContextMenu 
ArrowUp ArrowDown ArrowLeft ArrowRight
*/
const mod={
  //key to event attr
  "Shift":"shiftKey",
  "Control":"ctrlKey",
  "Alt":"altKey",
  "Meta":"metaKey"
};
const modrev={
  "shiftKey":"Shift",
  "ctrlKey":"Control",
  "altKey":"Alt",
  "metaKey":"Meta",
};
export function main(){
  const t=new Set();
  for(let e of k){
    if(e.key.length>1)t.add(e.key);
  }
  console.log(...t);
}
export const modifierState={
  shiftKey:false,
  ctrlKey:false,
  altKey:false,
  metaKey:false,
};
let target=new EventTarget();
export function push({type,key,code,keyCode}){
  let ko;
  if(key)ko=by.key.get(key);
  else if(keyCode)ko=by.keyCode.get(keyCode);
  else if(code)ko=by.code.get(code);
  if(!ko)ko={key,code,keyCode};
  
  const m=mod[ko.key];
  if(m)modifierState[m]=(type==="down");
  target.dispatchEvent(
    new KeyboardEvent(type,{
    ...ko,...modifierState
    }));
}
export function setTarget(t){
  target=t;
}
export function sync(to){
  // emit events that make modifierState to `to`.
  for(let k in to){
    if(modifierState[k] && !to[k]){
      push({type:"up",key:modrev[k]});
    }else if(!modifierState[k] && to[k]){
      push({type:"down",key:modrev[k]});
    }
  }
}
export function addEventListener(...a){
  return target.addEventListener(...a);
}
export function modifierStateToInt(e=modifierState){
      return 0 | 
    (e.ctrlKey ? 1 : 0) | 
    (e.altKey ? 2 : 0) | 
    (e.shiftKey ? 4 : 0) | 
    (e.metaKey ? 8 : 0);

}
