// @hoge1e3/global/global.js
// @hoge1e3/global
export function isPlainObject(o) {
  // do not use __proto__
  return (
    o !== null &&
    typeof o === "object" &&
    Object.getPrototypeOf(o) === Object.prototype
  );
}
const ids=["94956c36-a973-4b54-81c9-c053c20dfc4f",
"pNodeBootLoader"];
/** @type any */
const g=globalThis;
let value;
for(let id of ids){
  if(!g[id])continue;
  if(!value)value=g[id];
  else if(value!==g[id]){
    throw new Error("global Not match" +id);
  }
}
for(let id of ids){
  if(!g[id]){
    Object.defineProperty(
    g,id,{
      value
    });
  }
}
//@ts-check

/** @type any */
export function getGlobal() {
    return value;
}
/**
 * @param {string} k 
 * @returns any
 */
export function getValue(k) {
    return value[k] || g[k];
}
export const get=getValue;
export function set(k,v){
  assign({[k]:v});
}
/**
 * @param {object} o 
 */
export function pollute(o) {
    assign(o, value);
    assign(o, g);
}
/**
 * @param {any} o 
 */
export function assign(o, dst=value) {
    for (let k in o) {
        if (isPlainObject(o[k]) && isPlainObject(dst[k])) {
            assign(o[k], dst[k]);
        } else {
            dst[k]=o[k];
        }
    }
}
/**
 * @param {any} o 
 */
export function assignDefault(o, dst=value) {
    for (let k in o) {
        if (isPlainObject(o[k]) && isPlainObject(dst[k])) {
            assignDefault(o[k], dst[k]);
        } else {
            dst[k]=dst[k]??o[k];
        }
    }
}
assignDefault({
    getValue,
    assign, 
    assignDefault,
    pollute,
});
