// acepad/config.js
import {setConfig,getEditor} from './states.js';
//iport {toggleFullScreen} from './boot.js';
import {exec} from './command.js';
import {changeSession} from './sessions.js';
//imp {createFind} from './find.js';
import {unsetSel} from './editmode.js';
import {setModifier,renderModifierState} from './modifier.js';
//imprt {findSessionByFile} from './files.js';
//mport {showConsole} from './debug.js';
export function initConfig(){
    setConfig({
        commands:{
            "ctrl-w":(editor)=>editor.session.foldAll(),
            //"ctrl-q":()=>toggleFullScreen(),
            sessions:{bindKey:"ctrl-1",exec(){
                changeSession("*sessions*");
            }},
            quitEdit:{exec(editor){
                unsetSel(editor);
                editor.getSelection().clearSelection();
                setModifier("edit",0);
                renderModifierState();
            }},
        },
        menus:{
            /*
            home(){
                const editor=getEditor();
                const home=getHome();
                //todo  del 
                editor.setSession(findSessionByFile(home));
            },*/
        /*    full(){
                toggleFullScreen();
            },*/
            ses(){
                changeSession("*sessions*");
            },/*
            sh(){
                changeSession("*debug*");
            },*/
            gutt(){
                const editor=getEditor();
                editor.renderer.setShowGutter(
                    !editor.renderer.getShowGutter());
            },/*
            cons(){
                showConsole();
            },*/
        },
        jshint:{
            maxerr:10000,
            esnext:false,
            esversion:9,
            undef:true,
        }
    });
}
