// acepad/ngram.js
import {getCurrentEditor} from './states.js';
import {events} from './events.js';
import {getModifier} from './modifier.js';
import {textToButton,showGuide} from './keypad.js';
export function makeNgram(_n=4){
    n=_n;
    const c=new DefaultMap(()=>
    new Counter());
    ngram=c;
    events.on("renderModifierState",()=>{
        let e=getCurrentEditor();
        if(!e)return ;
        if(!e.session)return ;
        predict();
    });
    return c;
}
let n=4;
export var ngram;
export function addToNgram(s){
    for(let i=0;i<s.length-n;i++){
        const p=s.substring(i,i+n-1);
        const v=s[i+n-1];
        ngram.get(p).add(v);
    }
}
export function predict(session) {
    predict.remove();
    if(getModifier("ctrl"))return ;
    const editor=getCurrentEditor();
    session=session||editor.session;
    if (!ngram) return;
    const r=session.getSelection().getRange().start;
    const i=session.doc.positionToIndex(r);
    const pre=session.getValue().substring(i-(n-1),i);
    //console.log(pre,[...ngram.get(pre).keys()]);
    let c=textToButton();
    guides=sortValues(ngram.get(pre)).map((k,i)=>{
        if(!c[k])return ;
        i=10-i;
        if(i<=0)return ;
        return showGuide(c[k],i);
    }).filter(e=>e);
    return guides;
}
predict.remove=function (){
    for(let g of guides) g.remove();
};
export function sortValues(m){
    return [...m.entries()].sort(
        ([k,v],[k2,v2])=>v2-v
    ).map(([k,v])=>k);
}
export var guides=[];
export var DefaultMap=class extends Map{
    constructor(def){
        super();
        if(typeof def!=="function"){
            const v=def;
            def=()=>v;
        }
        this.def=def;
    }
    get(k){
        if(!this.has(k)){
            super.set(k,this.def(k));
        }
        return super.get(k);
    }
};
export var Counter=class extends DefaultMap{
    constructor(){
        super(0);
    }
    add(k){
        this.set(k,this.get(k)+1);
    }
};
