// @acepad/brg/js/brg.js
#!runts
import { c } from "@hoge1e3/component";
import { fromVal } from "@hoge1e3/immer-path-ref";
import { produce } from "immer";
export async function main() {
    await this.widget({
        wmain,
        css: "style.css"
    });
}
function parseDate(s) {
    const [y, m, d] = s.split("/").map(Number);
    return new Date(y, m - 1, d);
}
function formatDate(date) {
    const y = date.getFullYear();
    const m = String(date.getMonth() + 1).padStart(2, "0");
    const d = String(date.getDate()).padStart(2, "0");
    return `${y}/${m}/${d}`;
}
function todayKey() {
    return formatDate(new Date());
}
function rebuildCounts(dates) {
    const counts = { r: 0, g: 0, b: 0 };
    for (const { line } of dates) {
        for (const ch of line) {
            if (ch === "r" || ch === "g" || ch === "b")
                counts[ch]++;
        }
    }
    return counts;
}
function readLog(src) {
    /*
    src format
    2026/03/28 rb
    2026/03/30 g
    2026/03/31 r
    2026/04/01 rg
    2026/04/02 bb
    2026/04/03 r
    */
    const dates = src
        .split("\n")
        .map(l => l.trim())
        .filter(l => l.length > 0)
        .map(l => {
        const [dateStr, line = ""] = l.split(" ");
        return { date: parseDate(dateStr), line };
    });
    return { dates, counts: rebuildCounts(dates) };
}
function writeLog(log) {
    return log.dates
        .map(({ date, line }) => `${formatDate(date)} ${line}`)
        .join("\n");
}
function addm(log, e) {
    // date is current time, if exists append to it
    const key = todayKey();
    const existing = log.dates.find(d => formatDate(d.date) === key);
    if (existing) {
        existing.line += e;
    }
    else {
        log.dates.push({ date: new Date(), line: e });
    }
    log.counts[e]++;
}
function add(log, e) {
    return produce(log, (d) => addm(d, e));
}
async function wmain({ id: wid, t, on }) {
    // @acepad/brg
    const logf = this.resolve("log.txt");
    await this.edit(logf, { nomin: true });
    const logRef = fromVal(readLog(logf.text()));
    setTimeout(() => {
        console.log(document.getElementById(wid + "")?.innerHTML);
    }, 100);
    return c(wid.log, logRef, (log) => t.div(t.div({ id: wid, class: "bs" }, ...["b", "r", "g"].
        map(c => t.button({
        class: "btn " + c,
        onclick: () => {
            logRef.value = add(log, c);
            logf.text(writeLog(logRef.value));
        }
    }, log.counts[c]))), chiSquareBias(log)));
}
/**
 * Chi-squared goodness-of-fit test for uniform rgb distribution.
 * H0: r, g, b counts are equally distributed.
 * Returns p-value. Small p => biased.
 */
function chiSquareBias(log) {
    const vals = [log.counts.r, log.counts.g, log.counts.b];
    const total = vals.reduce((a, b) => a + b, 0);
    if (total === 0)
        return 1;
    const expected = total / 3;
    const chi2 = vals.reduce((a, b) => a + (b - expected) ** 2 / expected, 0);
    const df = 2; // k-1 = 3-1
    return pValueFromChi2(chi2, df);
}
/** p-value from chi-squared distribution (upper tail) */
function pValueFromChi2(chi2, df) {
    // P(X > chi2) = 1 - regularizedGammaP(df/2, chi2/2)
    return 1 - regularizedGammaP(df / 2, chi2 / 2);
}
/** Regularized lower incomplete gamma P(a, x) */
function regularizedGammaP(a, x) {
    if (x < 0)
        return 0;
    if (x === 0)
        return 0;
    if (x < a + 1)
        return gammaSeries(a, x);
    return 1 - gammaCF(a, x);
}
function gammaSeries(a, x) {
    const maxIter = 200, eps = 3e-7;
    let ap = a, sum = 1 / a, del = sum;
    for (let i = 0; i < maxIter; i++) {
        ap++;
        del *= x / ap;
        sum += del;
        if (Math.abs(del) < Math.abs(sum) * eps)
            break;
    }
    return sum * Math.exp(-x + a * Math.log(x) - lgamma(a));
}
function gammaCF(a, x) {
    const maxIter = 200, eps = 3e-7;
    let b = x + 1 - a, c = 1 / 1e-30, d = 1 / b, h = d;
    for (let i = 1; i <= maxIter; i++) {
        const an = -i * (i - a);
        b += 2;
        d = an * d + b;
        if (Math.abs(d) < 1e-30)
            d = 1e-30;
        c = b + an / c;
        if (Math.abs(c) < 1e-30)
            c = 1e-30;
        d = 1 / d;
        const del = d * c;
        h *= del;
        if (Math.abs(del - 1) < eps)
            break;
    }
    return Math.exp(-x + a * Math.log(x) - lgamma(a)) * h;
}
function lgamma(x) {
    // Stirling approximation
    const c = [76.18009172947146, -86.50532032941677, 24.01409824083091,
        -1.231739572450155, 0.1208650973866179e-2, -0.5395239384953e-5];
    let y = x, tmp = x + 5.5;
    tmp -= (x + 0.5) * Math.log(tmp);
    let ser = 1.000000000190015;
    for (const ci of c)
        ser += ci / ++y;
    return -tmp + Math.log(2.5066282746310005 * ser / x);
}
