// petit-node/petit-node.js
// petit-node
import * as p from "pnode:main";
export const CJSCompiler=p.CJSCompiler;
export const CompiledESModule=p.CompiledESModule;
export const ESModule=p.ESModule;
export const ESModuleCompiler=p.ESModuleCompiler;
export const FS=p.FS;
export const ModuleEntry=p.ModuleEntry;
export const NodeModule=p.NodeModule;
export const _import=p._import;
export const addAlias=p.addAlias;
export const addAliases=p.addAliases;
export const addPrecompiledCJSModule=p.addPrecompiledCJSModule;
export const addPrecompiledESModule=p.addPrecompiledESModule;
export const boot=p.boot;
export const convertStack=p.convertStack;
export const core=p.core;
export const createModuleURL=p.createModuleURL;
export default p.default;
export const events=p.events;
export const file=p.file;
export const getAliases=p.getAliases;
export const getCore=p.getCore;
export const getFS=p.getFS;
export const getNodeLikeFs=p.getNodeLikeFs;
export const importModule=p.importModule;
export const init=p.init;
export const loadedModules=p.loadedModules;
export const on=p.on;
export const require=p.require;
export const resolveEntry=p.resolveEntry;
export const thisUrl=p.thisUrl;
export const urlToFile=p.urlToFile;
export const urlToPath=p.urlToPath;
export const version=p.version;
export const definedInAcepad=true;
