// ace-overlay/ace-overlay.js
/*global ace*/

export function rect(e){
    const test=e;
    const inf={
        left: test.offsetLeft,
        top: test.offsetTop,
        width: test.offsetWidth,
        height: test.offsetHeight,
    };//getElementRelativeInfo(test);
    return inf;
}

export function overlay(e){
    const test=e;
    const cs=(getComputedStyle(test));
    const fsz=cs.getPropertyValue("font-size");
    
    const overp=test.offsetParent;//document.querySelector("#over");
    const over=document.createElement("div");
    const inf=rect(e);
    console.log("inf",inf);
    over.setAttribute("style",`
    position: absolute;
    left: ${inf.left}px;
    top: ${inf.top}px;
    width: ${inf.width}px;
    height: ${inf.height}px;
    `);
    overp.appendChild(over);
    const editor= ace.edit(over);
    editor.session.setValue(e.value);
    editor.setFontSize(fsz);
    editor.renderer.setShowGutter(false);
    editor.on("input",(ev)=>{
        e.value=editor.session.getValue();
        e.dispatchEvent(new Event("input"));
    });
    editor.on("blur",(ev)=>{
        e.value=editor.session.getValue();
        e.dispatchEvent(new Event("change"));
    });
    
    return editor;
}
