// @hoge1e3/ngramts/test/test-ngram.ts
import {Document,Index,DocumentSet, DocumentIndex, LetterIndex} from "../src/ngram.js";
import * as assert from "assert";
import {SFile} from "@hoge1e3/sfile";

function echo(...args:any[]){
    console.log(...args);
}
function collectWords(words: Set<string>, content: string) {
    // Split the content into words using a regex to handle different delimiters (e.g., spaces, punctuation).
    const extractedWords = content.split(/\W+/);
    // Iterate through the extracted words.
    for (const word of extractedWords) {
        // Add non-empty words to the set.
        if (word.trim()) {
            words.add(word.trim().toLowerCase()); // Convert to lowercase to avoid duplicates.
        }
    }
}
function getFixture(dir:SFile,ds:DocumentSet){
    const s=new Set<string>();
    for(let f of dir.recursive()){
        if(f.ext()!==".js")continue;
        const d=new Document(f.path(),f.text());
        ds.addDocument(d);
        collectWords(s,d.content);
    }
    return s;
}

export async function main(this:any){
    const fixture=this.rel("/jsmod/node_modules");
    let d=new Document("A","abcc deff adc eedef abc");
    let s=new DocumentSet();
    const words=getFixture(fixture, s);
    this.echo([...words].slice(0,10));
    s.addDocument(d);
    let scale="";
    for(let i=0;i<d.content.length;i++){
        scale+=(i%10);
    }
    echo(scale, d.content.length);
    echo(d.content);
    //traverse(s.root);
    /*for(let i=0;i<15;i++){
        echo(s.expand());
        s.calcCount();
        for(let word of allSubstr(d.content)){
            //echo(word);
            find(s,word);
        }
    }*/
    //traverse(s.root);
    //allWords(s.root);
    //await predict(s);
    //predictWord(s,d);
    //let {index,found,rest}=s.findIndex("bc");
    //this.echo(index.keyType,found,rest);
    //traverse.call(this,index);
    find(s,"ab");
    find(s,"bc");
    find(s,"def");
    //for(let s of allSubstr("abc"))echo(s);
}    
main();

function q(c:any){
    return `'${c}'`;
}
function allWords(idx: Index,ctx=""){
    if(idx instanceof DocumentIndex){
        for(let [doc, offsets] of idx.map){
            echo(q(ctx),doc,offsets.join(" "));
        }
    }else{
        for(let [chr, sidx] of idx.map){
            allWords(sidx,ctx+chr);
        }
    }
}
function* allSubstr(s:string){
    for(let i=0;i<s.length;i++){
        for(let j=i+1;j<=s.length;j++){
            yield s.substring(i,j);
        }
    }
}
function traverse(idx:Index){
    let p=(...a:any[])=>{
        echo(...a);
    };
    p("{");
    if (idx instanceof DocumentIndex) {
        for(let [k,v] of idx.map){
            p(q(k),v);
            p(",");
        }    
    } else {
        for(let [k,v] of idx.map){
            if(v instanceof LetterIndex){
                p(q(k)+": ");
                traverse(v);
            }else{
                p(k, v);
            }
            p(",");
        }        
    }
    p("}");
}
function find(ds:DocumentSet,word:string){
    let r=new Set();
    for(let {document,offset} of ds.find(word)){
        //echo(word,document.path,offset);
        assert.ok(document.looks(offset,word),
        [word,document.path,offset].join(","));
        r.add(`${document}:${offset}`);
    }
    for (let e of r) echo(word+": "+e);
    let offset=0, found=0;
    for(let [p,d] of ds.path2doc){
        while(true){
            let i=d.content.indexOf(word,offset);
            if(i<0)return ;
            offset=i;
            assert.ok(r.has(`${d}:${offset}`),
            `no ${d}:${offset}`);
            offset++;
            found++;
        }
    }
    assert.equal(r.size, found);
}
async function predict(ds:DocumentSet){
    for(let [p,d] of ds.path2doc){
        for(let i=1;i<d.content.length;i++){
            echo("predict",d.content.substring(i-5,i),"...");
            for(let [k,v] of await ds.predictLetter(d,i)){
                echo(`'${k}'(${v}) `,{n:1});
            }
            echo(" ");
        }
    }
}

function predictWord(ds:DocumentSet,d:Document){
    for(let w of ds.predictWord(d,1)){
        echo("pw",w.pre,w.post);
    }
}
