// @acepad/todo/todo.ts
// @acepad/todo
import {ref} from "@hoge1e3/ref";
import {id as idg} from "@hoge1e3/dom-id";
export interface Item{
  
}
export function decode(text:string){
  const lines = text.split(/\r?\n/).filter(l => l.trim().length > 0);
  const root = [];
  const stack = [{
    indent: -1, list: root }];
  for (let raw of lines) {
    // count leading spaces (indent)
    const indent = raw.match(/^ */)[0].length;
    let line = raw.trim();
    // detect "ok"
    let ok = false;
    if (line.startsWith("ok ")) {
      ok = true;
      line = line.slice(3);
    }
    const node = { 
      id:idg(),
      title: line };
    if (ok) node.ok = true;
    // find correct parent by popping stack
    while (stack.length > 1 && indent <= stack[stack.length - 1].indent) {
      stack.pop();
    }
    const parent = stack[stack.length - 1].list;
    parent.push(ref(node));
    // prepare for children
    node.sub = ref([]);
    stack.push({
      indent, 
      list: node.sub.value });
  }
  // optional: remove empty sub arrays
  /*function cleanup(list){
    for (const item of list) {
      if (item.sub && item.sub.length === 0) {
        delete item.sub;
      } else if (item.sub) {
        cleanup(item.sub);
      }
    }
  }
  cleanup(root);*/
  return ref(root);
}
// @acepad/todo
export function encode(data){
  const lines = [];

  function walk(list, depth){
    for (const itemr of list.value) {
      const item=itemr.value;
      const indent = " ".repeat(depth); // 1スペース単位
      const prefix = item.ok ? "ok " : "";
      lines.push(indent + prefix + item.title);
      
      if (item.sub && 
      item.sub.value.length > 0) {
        walk(item.sub, depth + 1);
      }
    }
  }

  walk(data, 0);
  return lines.join("\n");
}
