// @acepad/todo/todo.ts
// @acepad/todo
import {id as idg} from "@hoge1e3/dom-id";

export interface Item {
  id: string;
  title: string;
  ok?: boolean;
  sub: Item[];
}
export function findById(items:Item[], id:string):Item|null {
  const shallow=items.find((item)=>item.id===id);
  if (shallow) return shallow;
  for (let item of items) {
    const deep=findById(item.sub, id);
    if (deep) return deep;
  }
  return null;
}
export function findPathById(items: Item[], id: string): Item[] {
  for (let item of items) {
    if (item.id === id) {
      return [item];
    }
    const deepPath = findPathById(item.sub, id);
    if (deepPath.length > 0) {
      return [item, ...deepPath];
    }
  }
  return [];
}
export function sibling(
  itemRoot: Item[],
  id: string,
  offset: number): Item | null {
  const path = findPathById(itemRoot, id);
  if (path.length === 0) return null;

  const parent = path.length === 1
    ? itemRoot
    : path[path.length - 2].sub;

  const currentIndex = parent.findIndex((item) => item.id === id);
  if (currentIndex === -1) return null;

  const siblingIndex = currentIndex + offset;
  if (siblingIndex < 0 || siblingIndex >= parent.length) return null;
  return parent[siblingIndex];
}
export function decode(text:string):Item[]{
  const lines = text.split(/\r?\n/).filter(l => l.trim().length > 0);
  const root = [] as Item[];
  const stack = [{
    indent: -1, list: root }];
  for (let raw of lines) {
    // count leading spaces (indent)
    const indent = raw.match(/^ */)![0].length;
    let line = raw.trim();
    // detect "ok"
    let ok = false;
    if (line.startsWith("ok ")) {
      ok = true;
      line = line.slice(3);
    }
    const node:Item = { 
      id:idg().toString(),
      title: line,
      sub:[]
    };
    if (ok) node.ok = true;
    // find correct parent by popping stack
    while (stack.length > 1 && indent <= stack[stack.length - 1].indent) {
      stack.pop();
    }
    const parent = stack[stack.length - 1].list;
    parent.push(node);
    // prepare for children
    stack.push({
      indent, 
      list: node.sub });
  }
  // optional: remove empty sub arrays
  /*function cleanup(list){
    for (const item of list) {
      if (item.sub && item.sub.length === 0) {
        delete item.sub;
      } else if (item.sub) {
        cleanup(item.sub);
      }
    }
  }*/
  //cleanup(root);
  return (root);
}
// @acepad/todo
export function encode(data:Item[]){
  const lines = [] as string[];
  function walk(list:Item[], depth:number){
    for (const item of (list)) {
      const indent = " ".repeat(depth); // 1スペース単位
      const prefix = item.ok ? "ok " : "";
      lines.push(indent + prefix + item.title);
      
      if (item.sub && 
      item.sub.length > 0) {
        walk(item.sub, depth + 1);
      }
    }
  }
  walk(data, 0);
  return lines.join("\n");
}
