// @acepad/todo/ui.ts
#!runts
import * as kstr from "@hoge1e3/key-stream";
import {ReadonlyRef, ref,Ref} from "@hoge1e3/ref";
import {t,TagGenerator} from "@hoge1e3/dom";
import {decode, findById,Item} from "./todo.js";
import {Idgen} from "@hoge1e3/dom-id";
import {c} from "@hoge1e3/component";
import {refInput} from "@hoge1e3/ref-input";
import {create, PathRef, ReadonlyPathRef} from "@hoge1e3/immer-path-ref";
export async function main(this:any){
  return this.widget(wmain);
}    
function tap(a:any){
  console.log("tap",a)
  return a;
}
function wmain(this:any,{id:wid,t}:{id:Idgen, t:TagGenerator}){
  //this.echo("key");
  const root=decode(this.resolve(this.$home).rel("todo.txt").text());
  const expanded:Record<string,true>={};
  const expandedRef=create(ref(expanded));
  const rootRef=create(ref(root));
  const showok=ref(false)
  rootRef.on("change",({val})=>{
    console.log("r",val)
  })
  return t.div({style:`
  width:100%; height:100%; overflow: scroll
  `}, 
  refInput(showok,t.input({
    id:wid.so,
    type:"checkbox"
  })),
  todoListComponent(wid.root,rootRef));
  /*kstr.addEventListener("down",
  (e)=>this.echo(e.key))*/
  function todoListComponent(
    id:Idgen, items:PathRef<Item[]>) {
    return c(id, 
    items, showok,
    (items,so)=>t.ul(
      ...items.
      filter(i=>so||!i.ok).
      map(item=>todoComponent(
        rootRef.join((root)=>findById(root, item.id)!), 
        expandedRef.join(item.id))
    )))
  }
  function todoComponent(
    item: ReadonlyPathRef<Item>,
    expanded: PathRef<boolean>)
    :HTMLElement {
    const iid=wid[item.value.id]
    return c(iid, 
      item, expanded, 
      (itemv,expandedv)=>t.li(
          {onclick(e:any){
            console.log("ex",e.target)
            if(e?.target?.tagName.toLowerCase()!=="input")
            expanded.value=!expanded.value
          }},
        ...[itemv.sub.length?
          t.span(
          "+")
        :[]],
        
        refInput(item.join("ok"),
        t.input({id:iid.ok,type:"checkbox"})),
        refInput(item.join("title"),
        t.input({id:iid.title})),
        ...[expandedv?
          todoListComponent(
            iid.sub,
            item.join("sub"))
        :[]]
      ));
  }
}
