// @acepad/todo/ui.ts
#!runts
import * as kstr from "@hoge1e3/key-stream";
import {ReadonlyRef, ref,Ref,change,get,set} from "@hoge1e3/ref";
import {t,TagGenerator} from "@hoge1e3/dom";
import {decode, findById,Item,sibling} from "./todo.js";
import {Idgen} from "@hoge1e3/dom-id";
import {c} from "@hoge1e3/component";
import {refInput} from "@hoge1e3/ref-input";
import {fromVal,PathRef,ReadonlyPathRef} from "@hoge1e3/immer-path-ref";
import {prev,next} from "./nav.js"
//import {PathRefProxy,ReadonlyPathRefProxy,fromVal,map/*PathRef, ReadonlyPathRef*/} from "@hoge1e3/immer-path-ref/js/proxy.js";
export async function main(this:any){
  return this.widget({
    wmain,
    css:"todo.css"
  });
} 
function wmain(this:any,
{id:wid,t,on}:
{id:Idgen, t:TagGenerator,on:Function}){
  //this.echo("key");
  const root=decode(this.resolve(this.$home).rel("todo.txt").text());
  const expands:Record<string,boolean>={};
  const expandsRef=fromVal(expands);
  const rootRef=fromVal(root);
  const showok=ref(false)
  const activeRef=ref(root[0].id);
  change(rootRef,({val})=>{
    console.log("r",val)
  })
  on("close",()=>kstr.removeEventListener("keydown",kdh))
  kstr.addEventListener("keydown",kdh);
  function kdh(e:KeyboardEvent){
    const o={ArrowUp:-1,
    ArrowDown:1}[e.key]
    if(!o)return 
    const active=activeRef.value
    if(!active)return false;
    const n=next(
      document.querySelector(".todo.active")!,
    ".todo",o);
    console.log(n,active)
    if(!n)return false
    activeRef.value=n.getAttribute("data-id")!;
    setTimeout(()=>{
      document.querySelector(".todo.active .title")!.
      dispatchEvent(new Event("click"))
      
    },100)
    return true;
  }
  setTimeout(()=>
  console.log("q",document.querySelectorAll(".todo")),100)
  return t.div(
    {class:"scroll"},
    t.div(
      refInput(showok,t.input({
        id:`${wid}-showok`,
        type:"checkbox"
      })),"Show ok"),
    todoListComponent(`${wid}-root`,rootRef));
  
  function todoListComponent(
    id:string, items:PathRef<Item[]>) {
    return c(id, 
    items, showok,
    (items,so)=>t.ul(
      ...items.
      filter(i=>so||!i.ok).
      map(item=>todoComponent(
        rootRef.join((root)=>findById(root, item.id)!), 
        expandsRef.join(item.id))
    )))
  }

  function hasSub(i: Item, showok: boolean): boolean {
    if (showok) return i.sub.length > 0;
    return i.sub.some(child => !child.ok || hasSub(child, showok));
  }
  function op(c:boolean,f:()=>HTMLElement){
    return c?[f()]:[];
  }
  function todoComponent(
    itemRef: ReadonlyPathRef<Item>,
    expandedRef: PathRef<boolean>)
    :HTMLElement {
    const itemid=itemRef.join("id").value
    const iid=wid+"-"+itemid
    const isaRef=computed(activeRef,
    (a)=>a===itemid)
    return c(iid, 
      itemRef, expandedRef, activeRef,
      (item,expanded,active)=>t.li(
        {
          class: "todo "+
            (active===item.id?
            "active":""),
          "data-id":item.id,
          onclick(e:any){
            console.log("ex",e.target)
            if(e?.target?.tagName.toLowerCase()!=="input"){
              expandedRef.value=!expanded
            }
            activeRef.value=item.id
            e.stopPropagation();
          }
        },
        ...op(
          hasSub(item,showok.value),
          ()=>t.span("+")),
        
        refInput(itemRef.join("ok"),
        t.input({id:iid+"-ok",type:"checkbox"})),
        refInput(itemRef.join("title"),
        t.input({id:iid+"-title",class:"title"})),
        ...op(expanded,()=>
          todoListComponent(
            iid+"-sub", 
            itemRef.join("sub")))
      ));
  }
}
function tap(a:any){
  console.log("tap",a)
  return a;
}
