// @hoge1e3/dom/dom.js
// @hoge1e3/dom
/* global Reflect*/
import isPlainObject from "@hoge1e3/is-plain-object";
import EventManager from "@hoge1e3/dom-mutable-events";
export const sym_eventHandlerNames=Symbol("sym_eventHandlerNames");
const evm=new EventManager();
export function generator() {
    const prop={};
    let pr=new Proxy(prop,{
        get(o,key){
            if(key in prop || typeof key==="symbol"){
                return Reflect.get(prop,key);
            }
            return tagGenerator(key);
        },
        set(o,key,val){
            return Reflect.set(prop,key,val);
        },
        /*apply(o,self,args){
        }*/
    });
    return pr;
}
function isPrimitive(v) {
    return (
        typeof v!=="function" && 
        typeof v!=="object") || v==null;
}
export function tagGenerator(tag) {
    return (...args)=>{
        const d=document.createElement(tag);
        const add=(a)=>{
            if(isPlainObject(a)){
                for(let k in a){
                    if(typeof k==="symbol") {
                        d[k]=a[k];
                    } else if (k.match(/^on/)){
                        //d[k]=a[k];
                        evm.add(d, k.substring(2),a[k]);
                        //addEventHandlerName(d,k);
                    } else if(k==="style"&&typeof a[k]==="object"){
                        d.setAttribute(k,cssattr(a[k]));
                    }else{
                        d.setAttribute(k,a[k]);
                    }    
                }
            }else if(isPrimitive(a)){
                d.appendChild(document.createTextNode(a+""));
            }else if(Array.isArray(a)){
                while(a.length)add(a.shift());
            }else{
                d.appendChild(a);
            }
        };
        add(args);
        return d;
    };
}
/*
export function addEventHandlerName(d, name) {
    if (!d[sym_eventHandlerNames]){
        d[sym_eventHandlerNames]=new Set();
    }
    d[sym_eventHandlerNames].add(name);
}
export function getEventHandlerNames(d) {
    if (!d[sym_eventHandlerNames]){
        return new Set();
    }
    return d[sym_eventHandlerNames];
}*/
export function cssattr(a){
    return Object.keys(a).map(k=>`${k}: ${a[k]};`).
    join("\n");
}
export const t=generator();
